---
title: "Analysis of Paired-TAG H3K4me1 in the mouse brain"
date: "`r Sys.Date()`"
output: html_document
vignette: >
    %\VignetteIndexEntry{Example_PairedTag}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

## Introduction

In their recent [paper](https://doi.org/10.1038/s41592-021-01060-3), Zhu
et al. devellop the **Paired-Tag** method to jointly profile histone modifications
and transcriptome in single cells from the mouse brain. They profiles  active (H3K4me3,
H3K27ac, H3K36me3) and repressive (H3K27me3) histone marks as well as scRNA.

They showed that active histone mark H3k4me1 matched quite well the seperation
of mouse brain cell types observed in the corresponding transcriptomic data.
  
In this article we will show the results of a basic analysis of these data using
ChromSCape application.

## Data Access

Zhu et al. made the datasets visualizable & downloadable at this
[link](http://catlas.org/pairedTag) under the form of 
**Dense Matrix**.
  
Their data is also downloadable through GEO portal in the form
of **Sparse Matrix** at [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE152020).
The reads are counted on 5kbp genomic bins.


## Interactive analysis and visualization

The datasets were analysis interactively with ChromSCape and an archive containing
the datasets is available at [](). 
To vizualise the datasets directly in ChromSCape, unzip the directory and launch
ChromSCape :  


```{r, eval=TRUE, echo=TRUE}
library(ChromSCape)
library(gridExtra)
library(ggplot2)
```
```{r, eval=FALSE, echo=TRUE}
ChromSCape::launchApp()
```
  
  
Once the app has loaded, click on the '**Browse**' button under 
'**Select output directory**' select the root of the folder (e.g. the folder 
containing the 'ChromSCape_analyses' folder).  
The available datasets should now appear on the right box '**Current analysis**'.
Select the dataset you wish to vizualise/analyse and follow the flow of the app
by clicking on the left tabs that will be unlocked once the data has loaded.
 
**Note**:  The first time you'll open the application, you will be
guided through a small tour of the application, that you can come back to any
time you like by clicking the **Help** button on the upper right corner.  
Info buttons are also here to help you when you are not sure about some features
of the application.  
  
## H3K4me1 mark in the mouse brain

We will start with the repressive H3K27me3 histone mark. 

First we will load the pre-computed SingleCellExperiments (replace the "ChromSCape_path"
variable with the path where you downloaded the computed analysis).
  
  
```{r, eval=TRUE, echo=TRUE}
# Load the preprocessed data (in the 'Filtering_Normalize_Reduce' folder)
ChromScape_path = "/media/pprompsy/Depic_bioinfo_1/InstitutCurie/Documents/Data/ChromSCape_Data/ChromSCape_1.5.22/ChromSCape_analyses/"

scExp = qs::qread(file.path(ChromScape_path,"GSE152020_PairedTag_Zhu_2021_H3K4me1", "Filtering_Normalize_Reduce","GSE152020_PairedTag_Zhu_2021_H3K4me1_500_200000_99_uncorrected.qs"))
```
  
  
## PCA & UMAP plots of replicates & total counts
  
Let's first see how many cells are present in this dataset:


```{r, eval=TRUE, echo=TRUE}
# Visualize a table of each replicate
num_cell_scExp(as.data.frame(SingleCellExperiment::colData(scExp)),
               datamatrix = SingleCellExperiment::counts(scExp))
```
  
  
We can now plot the PCA and the UMAP of the dataset. Note
that the first Principal Component (PC) of the PCA was removed as with TF-IDF
normalization the first PC is almost always driven by count. We recommend using
TF-IDF normalization when choosing small genomic bins such as 5-kbp bins which
is the case in this analysis.  
  
This is why the first component of the PCA was removed from the analysis as it is 
almost always strongly driven by the total counts (library size).



```{r, eval=TRUE, echo=TRUE, fig.height=7}
# Plot PCA (remember PC1 was removed as in TF-IDF + PCA, PC1 is driven by total count) & UMAP
p1 = plot_reduced_dim_scExp(scExp, color_by = "total_counts", reduced_dim = "PCA",
                            downsample = 10e6, transparency = 0.2, size = 0.3)
p2 = plot_reduced_dim_scExp(scExp, color_by = "total_counts", reduced_dim = "UMAP",
                            downsample = 10e6, transparency = 0.2, size = 0.3)
gridExtra::grid.arrange(p1,p2)
```
  
  
We observe a high amount of clusters on the UMAP with some light library size 
gradient within each cluster.

******

Let's vizualise the PCA and UMAP of the normalized data.

## Features contributing the most to cell type differenciation

A simple way to find interesting features and genes is to retrieve the features
that contributed the most to the PCA and therefore to the cell type differences.
  
Let's plot the list of the top 15 most contributing features to the second and
third component of the PCA. Green bars show positive contribution and red bars
show negative contribution. Features that have both a positive or negative 
contribution are probably correlated, while genes having opposite contribution 
will most likely be anti-correlated.
   
Let's plot feature contribution to the second component :
  
```{r, eval=TRUE, echo=TRUE}
plot_most_contributing_features(scExp,component = "Component_2", n_top_bot = 15)
```
  
  
Let's plot feature contribution to the second component :
  
```{r, eval=TRUE, echo=TRUE}
plot_most_contributing_features(scExp,component = "Component_3", n_top_bot = 15)
```
 
 We observe many intergenic regions that are important in separating the cell 
 types, which is quite interesting considering H3K4me1 is known to accumulate 
 mostly around genes.  
 
## Marker Genes

We can now take the markers used in the paper for H3K27me3 and color the UMAP with
the loci signal at the gene's locations. Note that for repressive marks such as
H3K27me3, the signal is supposedly anti-correlated with expression, therefore
cells likely to express marker genes are cells lacking the signal.
  
For Rfx4 and Olig2, it seems that those genes are marked by H3K27me3 in all 
clusters but 2 and 3.  
   
  
```{r, eval=TRUE, echo=TRUE}
marker_genes = c("Rfx4","Olig2","Pdgfra","Nkx6-2", "Alx4", "Neurod2", "Gad1", "Foxf2","Cttn")  
p = list()
for(i in seq_along(marker_genes)){
  gene = marker_genes[i]
  p[[i]] = plot_reduced_dim_scExp(scExp, color_by = gene, reduced_dim = "UMAP",
                                  downsample = 10e6, transparency = 0.2,
                                  size = 0.05, max_distanceToTSS = 2000
                                  ) + theme(axis.title = element_blank(),
                                                     legend.position = "none") +
                                  ggtitle(gene)
}
```

```{r, eval=TRUE, echo=FALSE, fig.width= 15}
gridExtra::grid.arrange(grobs = p[1:3], ncol = 3, nrow = 1)
```

```{r, eval=TRUE, echo=FALSE, fig.width= 15}
gridExtra::grid.arrange(grobs = p[4:6], ncol = 3, nrow = 1)
```

```{r, eval=TRUE, echo=FALSE, fig.width= 15}
gridExtra::grid.arrange(grobs = p[7:9], ncol = 3, nrow = 1)
```


  
## Clustering and correlations

Let's check the cell-to-cell correlation by plotting a heatmap. We will check
if cells group by correlation accordingly how they group on the UMAP feature
space.
  
This also shows us if there are cells with very low correlation scores, that 
might be low covered cells or artifacts that we can later filter out using the
`?filter_correlated_cell_scExp` function. 
  

```{r, eval=TRUE, echo=TRUE}
plot_heatmap_scExp(scExp, downsample = 2000)
```


Let's pick 8 clusters using hierarchical clustering on the correlation matrix,
as shown above the heatmap.

```{r, eval=TRUE, echo=TRUE, fig.height= 7}
set.seed(47)
scExp = find_clusters_louvain_scExp(scExp, k = 10)
plot_heatmap_scExp(scExp, downsample = 2000)
```

  
Most clusters seem to correspond to highly correlated cells. The biggest cluster 
in blue seem to be less homogeneous than the other.
  
Let's color the cells by cluster to see if the correlation clusters match the 
clusters observed in the 2D UMAP space.  

```{r, eval=TRUE, echo=TRUE}
plot_reduced_dim_scExp(scExp, color_by = "cell_cluster", reduced_dim = "UMAP", downsample = 10e6, size = 0.1, transparency = 0.2)
```
 
Interestingly, cluster C6 regroups cells that are separated on the UMAP. That
might indicate that these cells are differentiation from cluster C1 to cluster
C6 (or the contrary).
  
In our hands, correlation clustering can sometimes give results that are 
not perfectly matching clusters observed in the UMAP, that would be called 
directly by algorithm such as Louvain clustering. 

Let's assign cell identity cell identity which was assigned in Bartosovic et al.
 
```{r, eval=TRUE, echo=TRUE, fig.width=12}
metadata = read.table("http://catlas.org/pairedTag/cellbrowser/Paired-tag/meta.tsv", header = TRUE, sep = "\t")
metadata_json = jsonlite::read_json("http://catlas.org/pairedTag/cellbrowser/Paired-tag/dataset.json")
colors = paste0("#", unlist(metadata_json$metaFields[[15]]$colors))
df = data.frame(cell_cluster = unlist(metadata_json$metaFields[[15]]$valCounts)[seq(1,44,2)], cell_cluster_color = colors)
length(intersect(metadata$Cell, scExp$barcode)) #1 cell not in metadata
ncol(scExp)
scExp$cell_cluster = metadata$Annotation[match(scExp$barcode, metadata$Cell)]
scExp = colors_scExp(scExp, annotCol = "cell_cluster",
                     color_by = "cell_cluster", color_df = df)

plot_reduced_dim_scExp(scExp,
                       color_by = "cell_cluster",
                       reduced_dim = "UMAP",
                       downsample = 10e6,
                       size = 0.4,
                       transparency = 0.2,
                       annotate_clusters = TRUE)

```
  
The labels assigned by the authors do match the clusters observed in the UMAP,
the main difference being that there seem to be a smooth transition from Cortical
Neurons (in blue) to Inhibitory neurons (pink) but not a smooth transition towards
Hippocampal neurons (green). The non-neuron cells (orange) are well separated 
from neuron cells.
  
Interestingly, the cells in the center that are a mix from multiple annotations
in Zhu et al, are all clustered by hierarchical clustering together and with
cells that are more distant (see UMAP before - cluster C3). 


## Intra and Inter-cluster correlation

A interesting metrics is the cell-to-cell correlation between clusters, as
it can indicates which cell types share a common epigenetic landscape and which
cell types are far from each other.

```{r, eval=TRUE, echo=TRUE, fig.width=15}
plot_intra_correlation_scExp(scExp, by = "cell_cluster")
```


## Session information

```{r, eval=TRUE}
sessionInfo()
```
