---
title: "Analysis of scCUT&TAG in the mouse brain in different histone marks"
date: "`r Sys.Date()`"
output: html_document
---

## Introduction

In their recent [paper](https://doi.org/10.1038/s41587-021-00869-9), Bartosovic
et al. applied **scCUT&TAG** to the mouse brain to profile active (H3K4me3,
H3K27ac, H3K36me3) and repressive (H3K27me3) histone marks by adapting CUT&TAG to the
droplet-based 10X Genomics platform.  

They showed that the oligodendrocyte lineage (OLG) cell type heterogeneity can
be resolved using these epigenetic marks with more (H3K4me3, H3K27me3) or less
(H3K36me3, H3K27ac) resolution.
  
  
In this article we will show the results of a basic analysis of these data using
ChromSCape application.

## Data Access

Bartosovic et al. made the datasets visualizable & downloadable at this
[link](https://mouse-brain-cutandtag.cells.ucsc.edu) under the form of 
**Dense Matrix**.
  
Their data is also downloadable through GEO portal in the form
of H5 Seurat files at [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157637).
As these files are not directly readable by ChromSCape, we recommend downloading
the data through the first link of manually formating the data by reading the 
H5 files in R (see [Saving and Loading Data from an h5Seurat File
](https://mojaveazure.github.io/seurat-disk/articles/h5Seurat-load.html)).


## Interactive analysis and visualization

The datasets were analysis interactively with ChromSCape and an archive containing
the datasets is available at [](). 
To vizualise the datasets directly in ChromSCape, unzip the directory and launch
ChromSCape :  


```{r, eval=TRUE, echo=TRUE}
library(ChromSCape)
library(gridExtra)
library(ggplot2)
```
```{r, eval=FALSE, echo=TRUE}
ChromSCape::launchApp()
```
  
  
Once the app has loaded, click on the '**Browse**' button under 
'**Select output directory**' select the root of the folder (e.g. the folder 
containing the 'ChromSCape_analyses' folder).  
The available datasets should now appear on the right box '**Current analysis**'.
Select the dataset you wish to vizualise/analyse and follow the flow of the app
by clicking on the left tabs that will be unlocked once the data has loaded.
 
**Note**:  The first time you'll open the application, you will be
guided through a small tour of the application, that you can come back to any
time you like by clicking the **Help** button on the upper right corner.  
Info buttons are also here to help you when you are not sure about some features
of the application.  
  
## H3K27me3 mark in the mouse brain

We will start with the repressive H3K27me3 histone mark. 

First we will load the pre-computed SingleCellExperiments (replace the "ChromSCape_path"
variable with the path where you downloaded the computed analysis).
  
  
```{r, eval=TRUE, echo=TRUE}
# Load the preprocessed data (in the 'Filtering_Normalize_Reduce' folder)
ChromScape_path = "/media/pacome/Depic_bioinfo_1/InstitutCurie/Documents/Data/ChromSCape_Data/ChromSCape_3.13_3/ChromSCape_analyses/"

scExp = qs::qread(file.path(ChromScape_path,"Bartosovic_scCutTag_H3K27me3", "Filtering_Normalize_Reduce","Bartosovic_scCutTag_H3K27me3_100_40000_99.88_uncorrected.qs"))
```

## PCA & UMAP plots of replicates & total counts
  
First we will manually add the replicate information present in the 'barcode' 
column of the SingleCellExperiment. Based on the paper, we can assign the 
conditions of each replicate : 


```{r, eval=TRUE, echo=TRUE}
# 4 replicates (N1, N2, N3, N4)
scExp$sample_id = gsub("H3K27me3_|_[ATCG]*|:.*|-.*","", scExp$barcode)
scExp$sample_id[which(scExp$sample_id == "N1")] = "N1_P15_GFP+"
scExp$sample_id[which(scExp$sample_id == "N2")] = "N2_P15_GFP+"
scExp$sample_id[which(scExp$sample_id == "N3")] = "N3_P25_GFP+"
scExp$sample_id[which(scExp$sample_id == "N4")] = "N4_P25_GFP-"

# Color by replicate
scExp = colors_scExp(scExp, annotCol = "sample_id")

# Visualize a table of each replicate
num_cell_scExp(as.data.frame(SingleCellExperiment::colData(scExp)),
               datamatrix = SingleCellExperiment::counts(scExp))
```
  
  
We can now plot the PCA and the UMAP of the dataset, colored by replicate. Note
that the first Principal Component (PC) of the PCA was removed as with TF-IDF
normalization the first PC is almost always driven by count.

```{r, eval=TRUE, echo=TRUE, fig.height=7}
# Plot PCA (remember PC1 was removed as in TF-IDF + PCA, PC1 is driven by total count) & UMAP
p1 = plot_reduced_dim_scExp(scExp, color_by = "sample_id", reduced_dim = "PCA",
                            downsample = 10e6, transparency = 0.2, size = 0.3)
p2 = plot_reduced_dim_scExp(scExp, color_by = "sample_id", reduced_dim = "UMAP",
                            downsample = 10e6, transparency = 0.2, size = 0.3)
gridExtra::grid.arrange(p1,p2)
```
  
  
We observe clear clusters on the UMAP, that happen to be more discontinous than
what was observed in the original paper. On the UMAP, it  seems that some
clusters are specific to one or two replicates. Let's check that later when we 
will choose the clusters. This might be because replicates have s1 & 2 are older mice while
replicate 4 is GFP+ and other clusters GFP-.

******

Let's vizualise the PCA and UMAP of the normalized data. For this dataset we used
the TF-IDF normalization, as the cells are binned on small genomic bins (5000bp).
  
This is why the first component of the PCA was removed from the analysis as it is 
almost always strongly driven by the total counts (library size).

Let's plot the UMAP colored by the number of reads in each cell,
e.g. 'total_counts'.
  
  
```{r, eval=TRUE, echo=TRUE}
plot_reduced_dim_scExp(scExp, color_by = "total_counts", reduced_dim = "UMAP",
                            downsample = 10e6, transparency = 0.2, size = 0.2)
```
  

## Features contributing the most to cell type differenciation

A simple way to find interesting features and genes is to retrieve the features
that contributed the most to the PCA and therefore to the cell type differences.
  
Let's plot the list of the top 15 most contributing features to the second and
third component of the PCA. Green bars show positive contribution and red bars
show negative contribution. Features that have both a positive or negative 
contribution are probably correlated, while genes having opposite contribution 
will most likely be anti-correlated.
   
Let's plot feature contribution to the second component :
  
```{r, eval=TRUE, echo=TRUE}
plot_most_contributing_features(scExp,component = "Component_2", n_top_bot = 15)
```
  
  
Let's plot feature contribution to the second component :
  
```{r, eval=TRUE, echo=TRUE}
plot_most_contributing_features(scExp,component = "Component_3", n_top_bot = 15)
```
 
 As we can see, many genes involved in brain development & nervous sytem cell types are 
 present:  Olig2,  Neurod2,  Nkx6-2, Eomes, Barhl1, Lhx1, Prdm8, Prdm12,
 Islr2...  
  
  
## Marker Genes

We can now take the markers used in the paper for H3K27me3 and color the UMAP with
the loci signal at the gene's locations. Note that for repressive marks such as
H3K27me3, the signal is supposedly anti-correlated with expression, therefore
cells likely to express marker genes are cells lacking the signal.
  
For Rfx4 and Olig2, it seems that those genes are marked by H3K27me3 in all 
clusters but 2 and 3.  
  
    
  
```{r, eval=TRUE, echo=TRUE}
OLG_marker_genes = c("Rfx4","Olig2","Pdgfra","Nkx6-2", "Alx4", "Neurod2", "Gad1", "Foxf2","Cttn") 
p = list()
for(i in seq_along(OLG_marker_genes)){
  gene = OLG_marker_genes[i]
  p[[i]] = plot_reduced_dim_scExp(scExp, color_by = gene, reduced_dim = "UMAP",
                                  downsample = 10e6, transparency = 0.2,
                                  size = 0.5, max_distanceToTSS = 15000
                                  ) + theme(axis.title = element_blank(),
                                                     legend.position = "none") +
                                  ggtitle(gene)
}

```

```{r, eval=TRUE, echo=FALSE, fig.width= 15}
gridExtra::grid.arrange(grobs = p[1:3], ncol = 3, nrow = 1)
```

```{r, eval=TRUE, echo=FALSE, fig.width= 15}
gridExtra::grid.arrange(grobs = p[4:6], ncol = 3, nrow = 1)
```

```{r, eval=TRUE, echo=FALSE, fig.width= 15}
gridExtra::grid.arrange(grobs = p[7:9], ncol = 3, nrow = 1)
```


  
## Clustering and correlations

Let's check the cell-to-cell correlation by plotting a heatmap. We will check
if cells group by correlation accordingly how they group on the UMAP feature
space.
  
This also shows us if there are cells with very low correlation scores, that 
might be low covered cells or artifacts that we can later filter out using the
`?filter_correlated_cell_scExp` function. 
  

```{r, eval=TRUE, echo=TRUE}
plot_heatmap_scExp(scExp, downsample = 3000)
```

On the heatmap we can distinguish between 7 and 9 red squares, representing 
groups of highly correlated cells, e.g. clusters. 

  
Some cells seem to have low correlation values with other cells, and group with 
the largest cell cluster, let's filter lowly correlated cells first and then 
define the clusters.


```{r, eval=TRUE, echo=TRUE}
scExp_cf = filter_correlated_cell_scExp(scExp,
                                        random_iter = 5, # for speed
                                        corr_threshold = 99,
                                        percent_correlation = 1)
cat("Filtered ", ncol(scExp) - ncol(scExp_cf), " lowly correlated cells.\n")
```
  
Let's pick 8 clusters using hierarchical clustering on the correlation matrix,
as shown above the heatmap.

```{r, eval=TRUE, echo=TRUE, fig.height= 7}
set.seed(48)
scExp_cf = choose_cluster_scExp(scExp_cf, nclust = 8)
pdf("~/UMAP_cluster.pdf")
plot_reduced_dim_scExp(scExp_cf, color_by = "cell_cluster", reduced_dim = "UMAP", downsample = 10e6, size = 0.7, transparency = 0.2) +
  theme(legend.position = "none")
dev.off()

plot_heatmap_scExp(scExp_cf)
```

  
Most clusters seem to correspond to highly correlated cells. The biggest cluster 
in blue seem to be less homogeneous than the other.
  
Let's color the cells by cluster to see if the correlation clusters match the 
clusters observed in the 2D UMAP space.  

```{r, eval=TRUE, echo=TRUE}
plot_reduced_dim_scExp(scExp_cf, color_by = "cell_cluster", reduced_dim = "UMAP", downsample = 10e6, size = 0.1, transparency = 0.2)
```
 
Interestingly, cluster C6 regroups cells that are separated on the UMAP. That
might indicate that these cells are differentiation from cluster C1 to cluster
C6 (or the contrary).
  
In our hands, correlation clustering can sometimes give results that are 
not perfectly matching clusters observed in the UMAP, that would be called 
directly by algorithm such as Louvain clustering. 

Let's assign cell identity cell identity which was assigned in Bartosovic et al.
 
```{r, eval=TRUE, echo=TRUE}
metadata = read.table("https://cells.ucsc.edu/mouse-brain-cutandtag/h3k27me3/bins/meta.tsv",
                      header = TRUE, sep = "\t")

length(intersect(metadata$Cell, scExp_cf$barcode)) #1 cell not in metadata
scExp_cf = scExp_cf[,which(scExp_cf$barcode %in% metadata$Cell)]
scExp_cf$cell_cluster = metadata$cell_type[match(scExp_cf$barcode, metadata$Cell)]
scExp_cf = colors_scExp(scExp_cf, annotCol = "cell_cluster") # Assign replicate to sample_id

pdf("~/UMAP_annot.png")
plot_reduced_dim_scExp(scExp_cf, color_by = "cell_cluster", reduced_dim = "UMAP", downsample = 10e6, size = 0.7, transparency = 0.2) +
  theme(legend.position = "none")
dev.off()


pdf("~/UMAP_annot.pdf")
plot_reduced_dim_scExp(scExp_cf, color_by = "cell_cluster", reduced_dim = "UMAP", downsample = 10e6, size = 0.7, transparency = 0.2) +
  theme(legend.position = "none")
dev.off()

```
  
The labels assigned by the authors do match the clusters observed in the UMAP,
but differ slightly from the hierarchical clustering clusters (C1 and C6 are 
a mix of Astrocytes and Neurons_3). 

We can now visualize a nicely formatted table with the exact number of cells
per cell cluster (label).  
  
```{r, eval=TRUE, echo=TRUE}
scExp_cf$sample_id = scExp_cf$sample_id # Assign replicate to sample_id
scExp_cf = colors_scExp(scExp_cf, annotCol = "sample_id")
num_cell_in_cluster_scExp(scExp_cf)
```


We can see that Neurons are present only in GFP+ 

## Intra and Inter-cluster correlation

A interesting metrics is the cell-to-cell correlation between clusters, as
it can indicates which cell types share a common epigentic landscape and which
cell types are far from each other.

```{r, eval=TRUE, echo=TRUE, fig.width=15}
plot_intra_correlation_scExp(scExp_cf, by = "cell_cluster")
```
We observe the lowest intra-correlation scores for clusters C1 and C6, showing 
that the H3K27me3 profile of cells in these clusters is more heterogeneous
than in other cells.

  
Let's now plot correlation between C1 and the rest:

```{r, eval=TRUE, echo=TRUE, fig.width=15}
plot_inter_correlation_scExp(scExp_cf, by = "cell_cluster", reference_group = "VLMC")
```

## Session information

```{r, eval=TRUE}
sessionInfo()
```
