<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysis of scCUT&amp;TAG in the mouse brain in different histone marks • ChromSCape</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of scCUT&amp;TAG in the mouse brain in different histone marks">
<meta property="og:description" content="ChromSCape">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ChromSCape</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/vignette.html">Get started</a>
</li>
<li>
  <a href="../articles/vignette.html">Examples</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette.html">ChromSCape overview</a>
    </li>
    <li>
      <a href="../articles/scCutTag_Bartosovic_H3K27me3.html">scCUT&amp;TAG in the mouse brain</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/vallotlab/ChromSCape/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="scCutTag_Bartosovic_H3K27me3_files/header-attrs-2.8/header-attrs.js"></script><script src="scCutTag_Bartosovic_H3K27me3_files/kePrint-0.0.1/kePrint.js"></script><link href="scCutTag_Bartosovic_H3K27me3_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analysis of scCUT&amp;TAG in the mouse brain in different histone marks</h1>
            
            <h4 class="date">2021-05-21</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/vallotlab/ChromSCape/blob/master/vignettes/scCutTag_Bartosovic_H3K27me3.Rmd"><code>vignettes/scCutTag_Bartosovic_H3K27me3.Rmd</code></a></small>
      <div class="hidden name"><code>scCutTag_Bartosovic_H3K27me3.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>In their recent <a href="https://doi.org/10.1038/s41587-021-00869-9">paper</a>, Bartosovic et al. applied <strong>scCUT&amp;TAG</strong> to the mouse brain to profile active (H3K4me3, H3K27ac, H3K36me3) and repressive (H3K27me3) histone marks by adapting CUT&amp;TAG to the droplet-based 10X Genomics platform.</p>
<p>They showed that the oligodendrocyte lineage (OLG) cell type heterogeneity can be resolved using these epigenetic marks with more (H3K4me3, H3K27me3) or less (H3K36me3, H3K27ac) resolution.</p>
<p>In this article we will show the results of a basic analysis of these data using ChromSCape application.</p>
</div>
<div id="data-access" class="section level2">
<h2 class="hasAnchor">
<a href="#data-access" class="anchor"></a>Data Access</h2>
<p>Bartosovic et al. made the datasets visualizable &amp; downloadable at this <a href="https://mouse-brain-cutandtag.cells.ucsc.edu">link</a> under the form of <strong>Dense Matrix</strong>.</p>
<p>Their data is also downloadable through GEO portal in the form of H5 Seurat files at <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157637">GEO</a>. As these files are not directly readable by ChromSCape, we recommend downloading the data through the first link of manually formating the data by reading the H5 files in R (see <a href="https://mojaveazure.github.io/seurat-disk/articles/h5Seurat-load.html">Saving and Loading Data from an h5Seurat File</a>).</p>
</div>
<div id="interactive-analysis-and-visualization" class="section level2">
<h2 class="hasAnchor">
<a href="#interactive-analysis-and-visualization" class="anchor"></a>Interactive analysis and visualization</h2>
<p>The datasets were analysis interactively with ChromSCape and an archive containing the datasets is available at <a href=""></a>. To vizualise the datasets directly in ChromSCape, unzip the directory and launch ChromSCape :</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vallotlab/ChromSCape">ChromSCape</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">gridExtra</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ChromSCape</span><span class="fu">::</span><span class="fu"><a href="../reference/launchApp.html">launchApp</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>Once the app has loaded, click on the ‘<strong>Browse</strong>’ button under ‘<strong>Select output directory</strong>’ select the root of the folder (e.g. the folder containing the ‘ChromSCape_analyses’ folder).<br>
The available datasets should now appear on the right box ‘<strong>Current analysis</strong>’. Select the dataset you wish to vizualise/analyse and follow the flow of the app by clicking on the left tabs that will be unlocked once the data has loaded.</p>
<p><strong>Note</strong>: The first time you’ll open the application, you will be guided through a small tour of the application, that you can come back to any time you like by clicking the <strong>Help</strong> button on the upper right corner.<br>
Info buttons are also here to help you when you are not sure about some features of the application.</p>
</div>
<div id="h3k27me3-mark-in-the-mouse-brain" class="section level2">
<h2 class="hasAnchor">
<a href="#h3k27me3-mark-in-the-mouse-brain" class="anchor"></a>H3K27me3 mark in the mouse brain</h2>
<p>We will start with the repressive H3K27me3 histone mark.</p>
<p>First we will load the pre-computed SingleCellExperiments (replace the path with the path where you downloaded the computed analysis)</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load the preprocessed data (in the 'Filtering_Normalize_Reduce' folder)</span>
<span class="va">ChromScape_path</span> <span class="op">=</span> <span class="st">"/media/pacome/LaCie/InstitutCurie/Documents/Data/ChromSCape_Data/ChromSCape_3.13_2/ChromSCape_analyses/"</span>

<span class="va">scExp</span> <span class="op">=</span> <span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html">qread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">ChromScape_path</span>,<span class="st">"Bartosovic_scCutTag_H3K27me3"</span>, <span class="st">"Filtering_Normalize_Reduce"</span>,<span class="st">"Bartosovic_scCutTag_H3K27me3_100_40000_99.88_uncorrected.qs"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="pca-umap-plots-of-replicates-total-counts" class="section level2">
<h2 class="hasAnchor">
<a href="#pca-umap-plots-of-replicates-total-counts" class="anchor"></a>PCA &amp; UMAP plots of replicates &amp; total counts</h2>
<p>First we will manually add the recplicate information present in the ‘barcode’ column of the SingleCellExperiment.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># 4 replicates (N1, N2, N3, N4)</span>
<span class="va">scExp</span><span class="op">$</span><span class="va">replicate</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"H3K27me3_|_[ATCG]*|:.*|-.*"</span>,<span class="st">""</span>, <span class="va">scExp</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span>

<span class="co"># Color by replicate</span>
<span class="va">scExp</span> <span class="op">=</span> <span class="fu"><a href="../reference/colors_scExp.html">colors_scExp</a></span><span class="op">(</span><span class="va">scExp</span>, annotCol <span class="op">=</span> <span class="st">"replicate"</span><span class="op">)</span></code></pre></div>
<p>We can now plot the PCA and the UMAP of the dataset, colored by replicate. Note that the first Principal Component (PC) of the PCA was removed as with TF-IDF normalization the first PC is almost always driven by count.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Plot PCA (remember PC1 was removed as in TF-IDF + PCA, PC1 is driven by total count) &amp; UMAP</span>
<span class="va">p1</span> <span class="op">=</span> <span class="fu"><a href="../reference/plot_reduced_dim_scExp.html">plot_reduced_dim_scExp</a></span><span class="op">(</span><span class="va">scExp</span>, color_by <span class="op">=</span> <span class="st">"replicate"</span>, reduced_dim <span class="op">=</span> <span class="st">"PCA"</span>,
                            downsample <span class="op">=</span> <span class="fl">10e6</span>, transparency <span class="op">=</span> <span class="fl">0.1</span>, size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">=</span> <span class="fu"><a href="../reference/plot_reduced_dim_scExp.html">plot_reduced_dim_scExp</a></span><span class="op">(</span><span class="va">scExp</span>, color_by <span class="op">=</span> <span class="st">"replicate"</span>, reduced_dim <span class="op">=</span> <span class="st">"UMAP"</span>,
                            downsample <span class="op">=</span> <span class="fl">10e6</span>, transparency <span class="op">=</span> <span class="fl">0.1</span>, size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>
<span class="fu">gridExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html">grid.arrange</a></span><span class="op">(</span><span class="va">p1</span>,<span class="va">p2</span><span class="op">)</span></code></pre></div>
<p><img src="scCutTag_Bartosovic_H3K27me3_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>We observe clear clusters on the UMAP, that happen to be more discontinous than what was observed in the original paper. On the UMAP, it seems that some clusters are specific to one or two replicates. Let’s check that later when we will choose the clusters. This might be because replicates have s1 &amp; 2 are older mice while replicate 4 is GFP+ and other clusters GFP-.</p>
<hr>
<p>Let’s vizualise the PCA and UMAP of the normalized data. For this dataset we used the TF-IDF normalization, as the cells are binned on small genomic bins (5000bp).</p>
<p>This is why the first component of the PCA was removed from the analysis as it is almost always strongly driven by the total counts (library size).</p>
<p>Let’s plot the UMAP colored by the number of reads in each cell, e.g. ‘total_counts’.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_reduced_dim_scExp.html">plot_reduced_dim_scExp</a></span><span class="op">(</span><span class="va">scExp</span>, color_by <span class="op">=</span> <span class="st">"total_counts"</span>, reduced_dim <span class="op">=</span> <span class="st">"UMAP"</span>,
                            downsample <span class="op">=</span> <span class="fl">10e6</span>, transparency <span class="op">=</span> <span class="fl">0.2</span>, size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></code></pre></div>
<p><img src="scCutTag_Bartosovic_H3K27me3_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
<div id="features-contributing-the-most-to-cell-type-differenciation" class="section level2">
<h2 class="hasAnchor">
<a href="#features-contributing-the-most-to-cell-type-differenciation" class="anchor"></a>Features contributing the most to cell type differenciation</h2>
<p>A simple way to find interesting features and genes is to retrieve the features that contributed the most to the PCA and therefore to the cell type differences.</p>
<p>Let’s plot the list of the top 15 most contributing features to the second and third component of the PCA. Green bars show positive contribution and red bars show negative contribution. Features that have both a positive or negative contribution are probably correlated, while genes having opposite contribution will most likely be anti-correlated.</p>
<p>Let’s plot feature contribution to the second component :</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_most_contributing_features.html">plot_most_contributing_features</a></span><span class="op">(</span><span class="va">scExp</span>,component <span class="op">=</span> <span class="st">"Component_2"</span>, n_top_bot <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></code></pre></div>
<p><img src="scCutTag_Bartosovic_H3K27me3_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>Let’s plot feature contribution to the second component :</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_most_contributing_features.html">plot_most_contributing_features</a></span><span class="op">(</span><span class="va">scExp</span>,component <span class="op">=</span> <span class="st">"Component_3"</span>, n_top_bot <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></code></pre></div>
<p><img src="scCutTag_Bartosovic_H3K27me3_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>As we can see, many genes involved in brain development &amp; nervous sytem cell types are present: Olig2, Neurod2, Nkx6-2, Eomes, Barhl1, Lhx1, Prdm8, Prdm12, Islr2…</p>
</div>
<div id="marker-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#marker-genes" class="anchor"></a>Marker Genes</h2>
<p>We can now take the markers used in the paper for H3K27me3 and color the UMAP with the loci signal at the gene’s locations. Note that for repressive marks such as H3K27me3, the signal is supposedly anti-correlated with expression, therefore cells likely to express marker genes are cells lacking the signal.</p>
<p>For Rfx4 and Olig2, it seems that those genes are marked by H3K27me3 in all clusters but 2 and 3.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">OLG_marker_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Rfx4"</span>,<span class="st">"Olig2"</span>,<span class="st">"Pdgfra"</span>,<span class="st">"Nkx6-2"</span>, <span class="st">"Alx4"</span>, <span class="st">"Neurod2"</span>, <span class="st">"Gad1"</span>, <span class="st">"Foxf2"</span>,<span class="st">"Cttn"</span><span class="op">)</span> 
<span class="va">p</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span><span class="op">(</span><span class="va">OLG_marker_genes</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
  <span class="va">gene</span> <span class="op">=</span> <span class="va">OLG_marker_genes</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>
  <span class="va">p</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="../reference/plot_reduced_dim_scExp.html">plot_reduced_dim_scExp</a></span><span class="op">(</span><span class="va">scExp</span>, color_by <span class="op">=</span> <span class="va">gene</span>, reduced_dim <span class="op">=</span> <span class="st">"UMAP"</span>,
                                  downsample <span class="op">=</span> <span class="fl">10e6</span>, transparency <span class="op">=</span> <span class="fl">0.2</span>,
                                  size <span class="op">=</span> <span class="fl">0.05</span>, max_distanceToTSS <span class="op">=</span> <span class="fl">2000</span>
                                  <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
                                                     legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span>
                                  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="va">gene</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><img src="scCutTag_Bartosovic_H3K27me3_files/figure-html/unnamed-chunk-10-1.png" width="1440"></p>
<p><img src="scCutTag_Bartosovic_H3K27me3_files/figure-html/unnamed-chunk-11-1.png" width="1440"></p>
<p><img src="scCutTag_Bartosovic_H3K27me3_files/figure-html/unnamed-chunk-12-1.png" width="1440"></p>
</div>
<div id="clustering-and-correlations" class="section level2">
<h2 class="hasAnchor">
<a href="#clustering-and-correlations" class="anchor"></a>Clustering and correlations</h2>
<p>Let’s check the cell-to-cell correlation by plotting a heatmap. We will check if cells group by correlation accordingly how they group on the UMAP feature space.</p>
<p>This also shows us if there are cells with very low correlation scores, that might be low covered cells or artifacts that we can later filter out using the <code><a href="../reference/filter_correlated_cell_scExp.html">?filter_correlated_cell_scExp</a></code> function.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_heatmap_scExp.html">plot_heatmap_scExp</a></span><span class="op">(</span><span class="va">scExp</span>, downsample <span class="op">=</span> <span class="fl">3000</span><span class="op">)</span></code></pre></div>
<p><img src="scCutTag_Bartosovic_H3K27me3_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>On the heatmap we can distinguish between 7 and 9 red squares, representing groups of highly correlated cells, e.g. clusters.</p>
<p>Some cells seem to have low correlation values with other cells, and group with the largest cell cluster, let’s filter lowly correlated cells first and then define the clusters.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scExp_cf</span> <span class="op">=</span> <span class="fu"><a href="../reference/filter_correlated_cell_scExp.html">filter_correlated_cell_scExp</a></span><span class="op">(</span><span class="va">scExp</span>,
                                        random_iter <span class="op">=</span> <span class="fl">5</span>, <span class="co"># for speed</span>
                                        corr_threshold <span class="op">=</span> <span class="fl">99</span>,
                                        percent_correlation <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<pre><code>## ChromSCape::filter_correlated_cell_scExp - Calculating correlation threshold using random matrix values...</code></pre>
<pre><code>## ChromSCape::filter_correlated_cell_scExp - Filtering low correlated cells...</code></pre>
<pre><code>## ChromSCape::filter_correlated_cell_scExp - Re-calculating hierarchical clustering...</code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"Filtered "</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">scExp</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">scExp_cf</span><span class="op">)</span>, <span class="st">" lowly correlated cells.\n"</span><span class="op">)</span></code></pre></div>
<pre><code>## Filtered  555  lowly correlated cells.</code></pre>
<p>Let’s pick 8 clusters using hierarchical clustering on the correlation matrix, as shown above the heatmap.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">47</span><span class="op">)</span>
<span class="va">scExp_cf</span> <span class="op">=</span> <span class="fu"><a href="../reference/choose_cluster_scExp.html">choose_cluster_scExp</a></span><span class="op">(</span><span class="va">scExp_cf</span>, nclust <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_heatmap_scExp.html">plot_heatmap_scExp</a></span><span class="op">(</span><span class="va">scExp_cf</span><span class="op">)</span></code></pre></div>
<p><img src="scCutTag_Bartosovic_H3K27me3_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>Most clusters seem to correspond to highly correlated cells. The biggest cluster in blue seem to be less homoegeneous than the other.</p>
<p>Let’s color the cells by cluster to see if the correlation clusters match the clusters observed in the 2D UMAP space.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_reduced_dim_scExp.html">plot_reduced_dim_scExp</a></span><span class="op">(</span><span class="va">scExp_cf</span>, color_by <span class="op">=</span> <span class="st">"cell_cluster"</span>, reduced_dim <span class="op">=</span> <span class="st">"UMAP"</span>, downsample <span class="op">=</span> <span class="fl">10e6</span>, size <span class="op">=</span> <span class="fl">0.1</span>, transparency <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></code></pre></div>
<p><img src="scCutTag_Bartosovic_H3K27me3_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>Interestingly, cluster C6 regroups cells that are separated on the UMAP. That might indicate that these cells are differentiation from cluster C1 to cluster C6 (or the contrary).</p>
<p>In our hands, correlation clustering can sometimes give results that are not perfectly matching clusters observed in the UMAP, that would be called directly by algorithm such as Louvain clustering.</p>
<p>Finally we can plot a nice table to see the repartition of replicates in each of the clusters.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scExp_cf</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">=</span> <span class="va">scExp_cf</span><span class="op">$</span><span class="va">replicate</span> <span class="co"># Assign replicate to sample_id</span>
<span class="va">scExp_cf</span> <span class="op">=</span> <span class="fu"><a href="../reference/colors_scExp.html">colors_scExp</a></span><span class="op">(</span><span class="va">scExp_cf</span>, annotCol <span class="op">=</span> <span class="st">"sample_id"</span><span class="op">)</span> <span class="co"># Assign replicate to sample_id</span>
<span class="fu"><a href="../reference/num_cell_in_cluster_scExp.html">num_cell_in_cluster_scExp</a></span><span class="op">(</span><span class="va">scExp_cf</span><span class="op">)</span></code></pre></div>
<div style="border: 1px solid #ddd; padding: 5px; ">
<table class="table table table-striped table-condensed"><tbody>
<tr>
<td style="text-align:center;">
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #4285F4 !important;">C1</span>
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #DB4437 !important;">C2</span>
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #F4B400 !important;">C3</span>
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #0F9D58 !important;">C4</span>
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #708090 !important;">C5</span>
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #000000 !important;">C6</span>
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #FFA500 !important;">C7</span>
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #00868B !important;">C8</span>
</td>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: black !important;">#Cells</span>
</td>
</tr>
<tr>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #4285F4 !important;">N1</span>
</td>
<td style="text-align:center;">
716
</td>
<td style="text-align:center;">
98
</td>
<td style="text-align:center;">
1057
</td>
<td style="text-align:center;">
313
</td>
<td style="text-align:center;">
356
</td>
<td style="text-align:center;">
100
</td>
<td style="text-align:center;">
12
</td>
<td style="text-align:center;">
38
</td>
<td style="text-align:center;">
2690
</td>
</tr>
<tr>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #DB4437 !important;">N2</span>
</td>
<td style="text-align:center;">
966
</td>
<td style="text-align:center;">
66
</td>
<td style="text-align:center;">
929
</td>
<td style="text-align:center;">
306
</td>
<td style="text-align:center;">
300
</td>
<td style="text-align:center;">
231
</td>
<td style="text-align:center;">
16
</td>
<td style="text-align:center;">
24
</td>
<td style="text-align:center;">
2838
</td>
</tr>
<tr>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #F4B400 !important;">N3</span>
</td>
<td style="text-align:center;">
1757
</td>
<td style="text-align:center;">
722
</td>
<td style="text-align:center;">
287
</td>
<td style="text-align:center;">
314
</td>
<td style="text-align:center;">
707
</td>
<td style="text-align:center;">
373
</td>
<td style="text-align:center;">
368
</td>
<td style="text-align:center;">
99
</td>
<td style="text-align:center;">
4627
</td>
</tr>
<tr>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #0F9D58 !important;">N4</span>
</td>
<td style="text-align:center;">
244
</td>
<td style="text-align:center;">
13
</td>
<td style="text-align:center;">
162
</td>
<td style="text-align:center;">
63
</td>
<td style="text-align:center;">
15
</td>
<td style="text-align:center;">
331
</td>
<td style="text-align:center;">
881
</td>
<td style="text-align:center;">
417
</td>
<td style="text-align:center;">
2126
</td>
</tr>
<tr>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: black !important;">#Cells</span>
</td>
<td style="text-align:center;">
3683
</td>
<td style="text-align:center;">
899
</td>
<td style="text-align:center;">
2435
</td>
<td style="text-align:center;">
996
</td>
<td style="text-align:center;">
1378
</td>
<td style="text-align:center;">
1035
</td>
<td style="text-align:center;">
1277
</td>
<td style="text-align:center;">
578
</td>
<td style="text-align:center;">
12281
</td>
</tr>
<tr>
<td style="text-align:center;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: black !important;">p-value</span>
</td>
<td style="text-align:center;">
&lt;0.00001
</td>
<td style="text-align:center;">
&lt;0.00001
</td>
<td style="text-align:center;">
&lt;0.00001
</td>
<td style="text-align:center;">
&lt;0.00001
</td>
<td style="text-align:center;">
</td>
<td style="text-align:center;">
&lt;0.00001
</td>
<td style="text-align:center;">
&lt;0.00001
</td>
<td style="text-align:center;">
&lt;0.00001
</td>
<td style="text-align:center;">
&lt;0.00001
</td>
</tr>
</tbody></table>
</div>
<p>Clusters 7 &amp; 8 seems to be enriched in N3 &amp; N4 populations.</p>
</div>
<div id="intra-and-inter-cluster-correlation" class="section level2">
<h2 class="hasAnchor">
<a href="#intra-and-inter-cluster-correlation" class="anchor"></a>Intra and Inter-cluster correlation</h2>
<p>A interesting metrics is the cell-to-cell correlation between clusters, as it can indicates which cell types share a common epigentic landscape and which cell types are far from each other.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_intra_correlation_scExp.html">plot_intra_correlation_scExp</a></span><span class="op">(</span><span class="va">scExp_cf</span>, by <span class="op">=</span> <span class="st">"cell_cluster"</span><span class="op">)</span></code></pre></div>
<p><img src="scCutTag_Bartosovic_H3K27me3_files/figure-html/unnamed-chunk-18-1.png" width="1440"> We observe the lowest intra-correlation scores for clusters C1 and C6, showing that the H3K27me3 profile of cells in these clusters is more heterogeneous than in other cells.</p>
<p>Let’s now plot correlation between C1 and the rest:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_inter_correlation_scExp.html">plot_inter_correlation_scExp</a></span><span class="op">(</span><span class="va">scExp_cf</span>, by <span class="op">=</span> <span class="st">"cell_cluster"</span>, reference_group <span class="op">=</span> <span class="st">"C1"</span><span class="op">)</span></code></pre></div>
<p><img src="scCutTag_Bartosovic_H3K27me3_files/figure-html/unnamed-chunk-19-1.png" width="1440"></p>
</div>
<div id="session-information" class="section level2">
<h2 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h2>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R Under development (unstable) (2021-04-08 r80148)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.5 LTS
## 
## Matrix products: default
## BLAS:   /home/pacome/R/R-devel/lib/libRblas.so
## LAPACK: /home/pacome/R/R-devel/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] ggplot2_3.3.3    gridExtra_2.3    ChromSCape_1.1.3
## 
## loaded via a namespace (and not attached):
##   [1] utf8_1.2.1                   shinydashboard_0.7.1        
##   [3] reticulate_1.20              tidyselect_1.1.1            
##   [5] RSQLite_2.2.7                AnnotationDbi_1.53.2        
##   [7] htmlwidgets_1.5.3            grid_4.1.0                  
##   [9] BiocParallel_1.25.5          Rtsne_0.15                  
##  [11] munsell_0.5.0                ScaledMatrix_0.99.2         
##  [13] ragg_1.1.2                   statmod_1.4.36              
##  [15] scran_1.19.23                DT_0.18                     
##  [17] umap_0.2.7.0                 miniUI_0.1.1.1              
##  [19] withr_2.4.2                  batchelor_1.7.15            
##  [21] colorspace_2.0-1             Biobase_2.51.0              
##  [23] filelock_1.0.2               highr_0.9                   
##  [25] knitr_1.33                   rstudioapi_0.13             
##  [27] stats4_4.1.0                 SingleCellExperiment_1.13.14
##  [29] labeling_0.4.2               MatrixGenerics_1.3.1        
##  [31] GenomeInfoDbData_1.2.6       farver_2.1.0                
##  [33] bit64_4.0.5                  rprojroot_2.0.2             
##  [35] vctrs_0.3.8                  generics_0.1.0              
##  [37] xfun_0.23                    BiocFileCache_1.99.8        
##  [39] rlist_0.4.6.1                R6_2.5.0                    
##  [41] GenomeInfoDb_1.27.13         ggbeeswarm_0.6.0            
##  [43] rsvd_1.0.5                   locfit_1.5-9.4              
##  [45] msigdbr_7.4.1                bitops_1.0-7                
##  [47] cachem_1.0.5                 DelayedArray_0.17.13        
##  [49] assertthat_0.2.1             promises_1.2.0.1            
##  [51] BiocIO_1.1.2                 shinycssloaders_1.0.0       
##  [53] scales_1.1.1                 beeswarm_0.3.1              
##  [55] gtable_0.3.0                 beachmat_2.7.9              
##  [57] qualV_0.3-3                  rlang_0.4.11                
##  [59] systemfonts_1.0.2            coop_0.6-2                  
##  [61] rtracklayer_1.51.5           lazyeval_0.2.2              
##  [63] yaml_2.2.1                   httpuv_1.6.1                
##  [65] tools_4.1.0                  ellipsis_0.3.2              
##  [67] kableExtra_1.3.4             jquerylib_0.1.4             
##  [69] BiocGenerics_0.37.6          Rcpp_1.0.6                  
##  [71] sparseMatrixStats_1.3.8      progress_1.2.2              
##  [73] zlibbioc_1.37.0              purrr_0.3.4                 
##  [75] RCurl_1.98-1.3               prettyunits_1.1.1           
##  [77] openssl_1.4.4                viridis_0.6.1               
##  [79] S4Vectors_0.29.19            zoo_1.8-9                   
##  [81] SummarizedExperiment_1.21.3  cluster_2.1.2               
##  [83] colorRamps_2.3               fs_1.5.0                    
##  [85] magrittr_2.0.1               data.table_1.14.0           
##  [87] RSpectra_0.16-0              ResidualMatrix_1.1.0        
##  [89] colourpicker_1.1.0           matrixStats_0.58.0          
##  [91] stringfish_0.15.1            hms_1.1.0                   
##  [93] qs_0.24.1                    shinyjs_2.0.0               
##  [95] mime_0.10                    evaluate_0.14               
##  [97] xtable_1.8-4                 XML_3.99-0.6                
##  [99] IRanges_2.25.11              compiler_4.1.0              
## [101] biomaRt_2.47.9               shinyhelper_0.3.2           
## [103] scater_1.19.13               tibble_3.1.2                
## [105] KernSmooth_2.23-20           crayon_1.4.1                
## [107] htmltools_0.5.1.1            later_1.2.0                 
## [109] tidyr_1.1.3                  RcppParallel_5.1.4          
## [111] DBI_1.1.1                    RApiSerialize_0.1.0         
## [113] dbplyr_2.1.1                 rappdirs_0.3.3              
## [115] babelgene_21.4               Matrix_1.3-3                
## [117] parallel_4.1.0               metapod_0.99.6              
## [119] igraph_1.2.6                 GenomicRanges_1.43.4        
## [121] forcats_0.5.1                pkgconfig_2.0.3             
## [123] pkgdown_1.6.1                GenomicAlignments_1.27.2    
## [125] plotly_4.9.3                 scuttle_1.1.20              
## [127] xml2_1.3.2                   svglite_2.0.0               
## [129] vipor_0.4.5                  bslib_0.2.5                 
## [131] dqrng_0.3.0                  stringdist_0.9.6.3          
## [133] webshot_0.5.2                XVector_0.31.1              
## [135] rvest_1.0.0                  stringr_1.4.0               
## [137] digest_0.6.27                ConsensusClusterPlus_1.55.0 
## [139] Biostrings_2.59.4            rmarkdown_2.8               
## [141] edgeR_3.33.7                 DelayedMatrixStats_1.13.6   
## [143] restfulr_0.0.13              curl_4.3.1                  
## [145] shiny_1.6.0                  Rsamtools_2.7.2             
## [147] rjson_0.2.20                 lifecycle_1.0.0             
## [149] jsonlite_1.7.2               BiocNeighbors_1.9.4         
## [151] askpass_1.1                  desc_1.3.0                  
## [153] viridisLite_0.4.0            limma_3.47.15               
## [155] fansi_0.4.2                  pillar_1.6.1                
## [157] lattice_0.20-44              shinyFiles_0.9.0            
## [159] KEGGREST_1.31.2              fastmap_1.1.0               
## [161] httr_1.4.2                   glue_1.4.2                  
## [163] png_0.1-7                    Sushi_1.29.0                
## [165] bluster_1.1.12               bit_4.0.4                   
## [167] stringi_1.6.2                sass_0.4.0                  
## [169] blob_1.2.1                   textshaping_0.3.4           
## [171] BiocSingular_1.7.2           memoise_2.0.0               
## [173] dplyr_1.0.6                  irlba_2.3.3</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Pacome Prompsy, Celine Vallot.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
