<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>ChromSCape • ChromSCape</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="ChromSCape">
<meta property="og:description" content="ChromSCape">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ChromSCape</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/vignette.html">Get started</a>
</li>
<li>
  <a href="../articles/vignette.html">Examples</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette.html">Vignettes overview</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/vallotlab/ChromSCape/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="vignette_files/header-attrs-2.8/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>ChromSCape</h1>
                        <h4 class="author">Pacome Prompsy, Pia Kirchmeier, Céline Vallot</h4>
            
            <h4 class="date">2021-05-18</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/vallotlab/ChromSCape/blob/master/vignettes/vignette.Rmd"><code>vignettes/vignette.Rmd</code></a></small>
      <div class="hidden name"><code>vignette.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Chromatin modifications orchestrate the dynamic regulation of gene expression during development and in disease. Bulk approaches have characterized the wide repertoire of histone modifications across cell types, detailing their role in shaping cell identity. However, these population-based methods do not capture <strong>cell-to-cell heterogeneity</strong> of chromatin landscapes, limiting our appreciation of the role of chromatin in dynamic biological processes. Recent technological developments enable the mapping of histone marks at single-cell resolution, opening up perspectives to characterize the heterogeneity of chromatin marks in complex biological systems over time. Yet, existing tools used to analyze bulk histone modifications profiles are not fit for the low coverage and sparsity of single-cell epigenomic datasets.</p>
<div class="figure">
<img src="../inst/www/ChromSCape_concept.png" alt=""><p class="caption"><br><br><strong>ChromSCape Overview: </strong> The application takes as input various single-cell format and let the user filter &amp; cluster cells, run differential analysis &amp; gene set enrichment analysis between epigenomic subpopulations, in an unsupervised manner.</p>
</div>
<p><strong>ChromSCape</strong> is a user-friendly interactive <strong>Shiny/R application</strong> that processes single-cell epigenomic data to assist the biological interpretation of chromatin landscapes within cell populations. <strong>ChromSCape</strong> analyses the distribution of repressive and active histone modifications as well as chromatin accessibility landscapes from single-cell datasets (<em>scATAC-seq, scChIP-seq, scCUT&amp;TAG…</em>).</p>
<p>The goal of <strong>ChromSCape</strong> is to provide user a interactive interface to explore and run a <strong>complete analyse</strong> (<em>QC, preprocessing, analysis, interpretation</em>) on various single-cell epigenomic data. The application has multiple advantages:</p>
<ul>
<li>No R knowledge is required</li>
<li>Analyses persist on the user’s computer, allowing for fast reloading of the analysis</li>
<li>Parameters are highly customizable</li>
<li>Vizualisation at all steps</li>
<li>Meant for multiple single cell epigenomic data</li>
<li>Acceptance of raw files (BED, BAMs, .txt, …) for feature counting</li>
<li>Linking epigenomics with biological interpretation with Gene Set Analysis</li>
</ul>
</div>
<div id="quick-start" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick Start</h1>
<p>To launch the application simply run:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"ChromSCape"</span><span class="op">)</span></code></pre></div>
<p>Load <strong>ChromSCape</strong></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Loading ChromSCape"</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "Loading ChromSCape"</code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vallotlab/ChromSCape">ChromSCape</a></span><span class="op">)</span></code></pre></div>
<p>Launch the shiny application</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/launchApp.html">launchApp</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>A browser should open with the application. If the browser doesn’t open automatically, navigate to the displayed url, e.g. “<em>Listening on <a href="http://127.0.0.1:5139" class="uri">http://127.0.0.1:5139</a></em>”. The first time you’ll open the application, you will be guided through a small tour of the application, that you can come back to any time you like by clicking the <strong>Help</strong> button on the upper right corner.</p>
</div>
<div id="chromscape-step-by-step" class="section level1">
<h1 class="hasAnchor">
<a href="#chromscape-step-by-step" class="anchor"></a>ChromSCape step by step</h1>
<p>The user interface is organized by ‘Tab’ corresponding to specific ‘step’ of the analysis. In order to be able to access to each Tab you need to complete the previous steps. For example, before accessing the ‘Filter &amp; Normalize’ Tab, you need to first complete the ‘Select &amp; Import’ Tab.<br>
Each one of your project is named an ‘Analysis’ and is comprised of one raw dataset and additional objects that you have produced by going further into the analysis. A folder named ‘<em>ChromSCape_analyses</em>’ which will contain all your Analysis is produced in the output directory when you create an Analysis for the first time. In this folder, each of your Analysis is a folder with the following organisation:</p>
<blockquote>
<p>ChromSCape_analyses/<br>
├── Analysis_1<br>
│   ├── annotation.txt<br>
│   ├── Filtering_Normalize_Reduce<br>
│   │   └── Analysis_1_(…).RData<br>
│   ├── correlation_clustering<br>
│   │   └── Analysis_1_(…).RData<br>
│   ├── Diff_Analysis_Gene_Sets<br>
│   │   └── Analysis_1_(…).RData<br>
│   └── scChIP_raw.RData</p>
</blockquote>
<p>The raw data is stored at the root of the folder, and at each main step (‘Filtering &amp; Normalization’, ‘Correlation Clustering’ and ‘Differential Analysis’) the objects are saved. This enable you to close the application and later re-load your analysis without the need of re-doing all those steps. This also enable you to share your analysis with colleagues simply by copying your Analysis folder.<br><em>Note:</em> The (…) in the saved objects contained the values of the parameters for an Analysis. If you try multiple parameter, each results will be saved this way and all trials will accessible in the future.</p>
<div id="input-files-before-launching-chromscape" class="section level2">
<h2 class="hasAnchor">
<a href="#input-files-before-launching-chromscape" class="anchor"></a>Input files (before launching ChromSCape)</h2>
<p>Various existing technologies allow to produce single-cell genome-wide epigenomic datasets :</p>
<ul>
<li>scChIP-seq (Grosselin et al., 2019) scCUT&amp;Tag (Kaya-Okur et al., 2019), scChIL-seq (Harada et al., 2019), scChIC-seq (Ku et al.,) reveal the distribution of histone marks (H3K27me3, H3K4me3) or transcription factors (RNA Polymerase 2,…) with single-cell resolution.</li>
<li>scATAC-seq (Buenorostro et al., 2015) or sciATAC-seq (Cusanovitch et al.,</li>
</ul>
<ol start="2016" style="list-style-type: decimal">
<li>assess regions of open chromatin in single-cells</li>
</ol>
<ul>
<li>scDNA methylation profiling, scRRBS, scWGBS … <br>
</li>
</ul>
<p><br> After sequencing a single-cell epigenomic experiment, the raw sequencing files (.fastq) are demultiplexed, aligned against a reference genome to output different kind of data depending on the technology. ChromSCape allows user to input a variety of different format. Depending on the output of the data-engineering/pre-processing pipeline used, the signal can be already <em>summarized into features</em> (<strong>Count matrix</strong>, <strong>Peak-Index-Barcode</strong> files) or stored directly in <em>raw format</em> (<strong>single-cell BAM</strong> or <strong>single-cell BED</strong> files).</p>
<p>Anyhow the format, ChromSCape needs signal to be summarized into features. If inputting <em>raw signal</em> (<strong>scBAM</strong> or <strong>scBED</strong>), the application lets user summarize signal of each cells into various features:</p>
<ul>
<li>
<strong>Genomic features</strong> (extended region around TSS of genes, enhancers)</li>
<li>
<strong>Peaks</strong> called on bulk or single-cell signal (BED file must be provided by the user)</li>
<li>
<strong>Genomic bins</strong> (windows of constant length, e.g. 100kbp, 50kbp, 5kbp…)</li>
</ul>
<p>Note that summarizing will take longer if using BAM files than BED files, and if the number of features is important (e.g. 5kbp bins, enhancers…).</p>
<div id="count-matrices-files" class="section level3">
<h3 class="hasAnchor">
<a href="#count-matrices-files" class="anchor"></a>Count matrices files</h3>
<p>Each sample should be contained into single-cell count matrix (features x cells) in tab-separated format (extension .txt or .tsv) or comma-separated format (.csv). The first column is genomic location in standard format (chr:start-end) and the next columns are reads counted in each cells in the corresponding genomic feature. Note that the first entry (row 1, column 1) must be empty. All files should be placed in the same folder.<br>
If inputing only a single matrix regrouping different samples, the user can check the ‘multiple sample’ check box and specify a number of samples. ChromSCape will automatically find the different samples based on the names of the cells, so please make sure samples names are all quite different (e.g. K562_.., GM12878_..).</p>
<p>An example of such dataset is availablefor H3K27me3 mouse scChIP-seq of paired PDX samples at: <a href="https://ndownloader.figshare.com/files/22629317">PDX mouse cells H3K27me3 scChIP-seq matrices</a>. <br></p>
<blockquote>
<p>BC969404 BC893525 BC239068 BC073314 chr10:0-50000 0 0 0 0 chr10:50000-100000 0 0 0 0 chr10:100000-150000 0 0 0 0 …</p>
</blockquote>
<p><br></p>
</div>
<div id="peak-index-barcode-files" class="section level3">
<h3 class="hasAnchor">
<a href="#peak-index-barcode-files" class="anchor"></a>Peak-Index-Barcode files</h3>
<p>This format regroups three files containing signal of all samples of one or multiple experiment:</p>
<ul>
<li>The <strong>barcode file</strong> contains one cell barcode per line and finish by ’_barcode.txt’ :</li>
</ul>
<blockquote>
<p>HBCx95_BC969404 HBCx95_BC893525 HBCx22_BC239068 HBCx22_BC073314 …</p>
</blockquote>
<ul>
<li>The <strong>peak file</strong> contains feature location (usually peaks) must be in BED format and finish by ’_peaks.bed’:</li>
</ul>
<blockquote>
<p>chr3 197959001 197961500 chr3 198080001 198081500 chr4 53001 55500 …</p>
</blockquote>
<ul>
<li>The <strong>index file</strong> contains indexes of non-zero signal and finish by ’_indexes.txt’, the first column is the index of peaks, the second column contains barcode index, the last column contains the signal:</li>
</ul>
<blockquote>
<p>459 1 1 461 1 1 556 1 2</p>
</blockquote>
<p><br></p>
</div>
<div id="single-cell-bam-or-bed-files" class="section level3">
<h3 class="hasAnchor">
<a href="#single-cell-bam-or-bed-files" class="anchor"></a>Single-cell BAM or BED files</h3>
<p>Each BAM or BED (.bed or .bed.gz) file must be grouping signal of a single-cell, and all files must be placed in the same folder. The signal will be summarized into either bins, peaks, or around gene TSS depending on user choice.</p>
<p>An example of such dataset is publicly available for H3K27me3 single-cell CUT&amp;TAG (Kaya-Okur et al., 2019) K562 and H1 cells at: <a href="https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE124680&amp;format=file">K562 H3K27me3 cells</a> and <a href="https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE129119&amp;format=file">K562 + H1 H3K27me3 cells</a>. All files need to be placed in the same folders, the BED files do not need to be unzipped.<br><br></p>
</div>
<div id="alignment-bam-files-for-peak-calling-optional" class="section level3">
<h3 class="hasAnchor">
<a href="#alignment-bam-files-for-peak-calling-optional" class="anchor"></a>Alignment BAM files for Peak Calling (<strong>optional</strong>)</h3>
<p>For the optional step of Peak Calling on cluster of cells found de novo, users have to input one BAM file per sample, placed in the same folder. The barcode information of each read should be contained in a specific tag (XB, CB..) and correspond to the column names of the corresponding count matrix.</p>
<p><br></p>
<blockquote>
<p>samtools view example_matrix.bam | head NS500388:436:HNG5VAFXX:2:21311:26430:3816 89 chr10 3102405 255 51M * 0 0 CTTGGTGTCTAGTGGATCTGCTGCAGTCTTCTGTTGTCAGTGCTAAATCAC EEEEE/E6EEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEAAAAA NH:i:1 HI:i:1 AS:i:50 nM:i:0 XS:i:2147483647 XD:Z:GATGACAAAG XB:Z:BC969404 …</p>
</blockquote>
</div>
</div>
<div id="create-import-a-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#create-import-a-dataset" class="anchor"></a>Create &amp; import a dataset</h2>
<p>Once you have launched the application, you arrive on the “<strong>Select &amp; Import</strong>”. Here you can selet an output directory, where your analyses will be saved. ChromSCape will automatically save the folder’s location so that you don’t have to select it each time you connect.</p>
<p>You must then name your analysis. The name shouldn’t contain special characters (except ’_‘). Choose either the Human (hg38) or Mouse (mm10) genome. This is used to annotate your features with the closest genes TSS for the Gene Set Analysis. Browse your computer for one or multiple matrices that will be analyzed together. To select multiple matrices, they must be placed in the same folder and then the user can select multiples matrices with (Shift + Click) or (Ctrl + Click). Finally, clicking on ’Create Analysis’ will create an analysis &amp; import the count matrices in this analysis. This will create a folder named ‘<em>ChromSCape_analyses</em>’ in the output directory you specified, inside which another folder ‘Your_analysis_name’ is nested. If you create another analysis, it will also be created under ‘<em>ChromSCape_analyses</em>’. If you already created an analysis in a previous section, selecting the same output directory as you chose in the previous session will enable you to load your analysis.</p>
<p><br></p>
<div class="figure">
<img src="../inst/www/Guide_images/Step1_select_import.png" alt=""><p class="caption"><br><br><strong>Select &amp; Import Tab: </strong> Select an output directory, Create an analysis or load previous analyses, import count matrix into your new analysis.</p>
</div>
<div class="figure">
<img src="../inst/www/Guide_images/Step1_format_input.png" alt=""><p class="caption"><br><br> Select an type of format input, and eventually a type of feature to summarize signal.</p>
</div>
<div id="summarizing-features" class="section level3">
<h3 class="hasAnchor">
<a href="#summarizing-features" class="anchor"></a>Summarizing features</h3>
<p>Depending on what kind of data you are analyzing, you might want to summarized your data on different features. For open/active chromatin measures, such as ATAC or ChIP against H3K4me3, we recommand to summarize signal into small bins of 5kbp to 10kbp or to count aroung gene TSS. When counting around gene TSS, the default is to take overlapping regions of 2,500bp around the gene TSS, but you might want to increase this value for scATAC-seq if looking to retrieve enhancers open regions. For repressive marks such as H3K27me3, we recommand summarizing into large bins of 50kbp to 100kbp, as those mark are known to accumulate into large domains.</p>
</div>
<div id="recognition-of-samples" class="section level3">
<h3 class="hasAnchor">
<a href="#recognition-of-samples" class="anchor"></a>Recognition of samples</h3>
<p>ChromSCape will recognize automatically to which sample/condition each cell belongs. In order to do so, the names of the cells (e.g. names of the files for scBED/scBAM or barcodes names for combined Index-Peaks-Barcode/count matrix) should explicitely contains name of the samples/conditions.</p>
<p>Example: if inputing two cell types, e.g. Jurkat and Ramos, names of the files/cells should be something like : Jurkat_cell1.BED, Jurkat_cell2.BED, …, Ramos_cell1.BED, Ramos_cell2.BED.</p>
<p>The more different the names of the samples/conditions are, the more easily ChromSCape will classify correctly each cell. If having two samples with a similar name, users should manually change the names of the label. For example if having ‘CMP’ (common myeloid progenitors) and ‘GMP’ (granulocytes myeloid progenitors) samples, change the name manually to ‘commonMP’ and ‘granulocytesMP’ respectively.</p>
</div>
<div id="inputing-bam-files" class="section level3">
<h3 class="hasAnchor">
<a href="#inputing-bam-files" class="anchor"></a>Inputing BAM files</h3>
<p>Inputting BAM files is the most demanding task in terms of ressource in the application, despite the fact that the summarizing of features will take place in parallel. To give an example, for 1,287 single-cell BAM files, summarizing into 50kbp bins takes 9 minutes to finish. Counting from BED files is way faster, therefore if analyzing lots of cells in BAM format, users might consider running ‘bamToBed’ command from BEDTools on each file beforehand.</p>
</div>
</div>
<div id="filter-normalize-reduce-dimensionality" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-normalize-reduce-dimensionality" class="anchor"></a>2. Filter, Normalize &amp; Reduce Dimensionality</h2>
<p>In order to efficiently remove outlier cells from the analysis, e.g. cells with excessively high or low coverage, the user sets a threshold on a minimum read count per cell and the upper percentile of cells to remove. The latter could correspond to doublets, e.g. two cells in one droplet, while lowly covered cells are not informative enough or may correspond to barcodes ligated to contaminant DNA or library artifacts. Regions not supported by a minimum user-defined percentage of cells that have a coverage greater than 1,000 reads are filtered out.</p>
<p>Defaults parameters were chosen based on the analysis of multiple scChIP-seq datasets:</p>
<ul>
<li>A minimum coverage of 1,600 unique reads per cell</li>
<li>Filtering out the cells with the top 5% coverage</li>
<li>Keeping regions detected in at least 1 % of cells.</li>
</ul>
<p>Post quality control filtering, the matrices are normalized by total read count and region size (<a href="https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/">CPM normalization</a>. At this step, the user can provide a list of genomic regions, in BED format, to exclude from the subsequent analysis, in cases of known copy number variation regions between cells for example.</p>
<p>To reduce the dimensions of the normalized matrix for further analysis, PCA is applied to the matrix, with centering, and the 50 first PCs are kept for further analysis.</p>
<p>If you want to run t-SNE for vizualisation in addition to PC and UMAP, tick the ‘Run T-SNE’ checkbox. This might take some time to run, depending on the size of the dataset.</p>
<p>For some large datasets (e.g. more than 15,000 cells) or unbalance datasets (i.e. number of cells per sample is very different), it might be a good option to first run with the subsampling option. To do so, tick the ‘Perform subsampling’ and select the maximum number of cells to subsample from each sample. The sampling is done without replacement.</p>
<p>Users might want to remove specific loci from the analysis, e.g. repeat regions, known Copy Number Altered regions that might bias the analysis. To do so, tick the ‘Exclude specific genomic regions’ and browse for a BED file containing the regions to exclude.</p>
<p>A batch correction option using mutual nearest neighbors ‘FastMNN’ function from <a href="https://www.bioconductor.org/packages/release/bioc/html/batchelor.html">batchelor</a> is implemented to remove any known batch effect in the reduced feature space. To do so, tick the ‘Perform batch correction’, select the number of batches and add each sample to a batch.</p>
<p><br></p>
<div class="figure">
<img src="../inst/www/Guide_images/Step2_filter_normalize.png" alt=""><p class="caption"><br><br><strong>Filter, Normalize &amp; Reduce Tab: </strong> Adjust paramters to filter out low covered cells &amp; features, as well as too highly covered cells that might be potential doublets. Normalize &amp; Reduce dimensionality using PCA.</p>
</div>
</div>
<div id="visualize-cells-in-reduced-dimensions" class="section level2">
<h2 class="hasAnchor">
<a href="#visualize-cells-in-reduced-dimensions" class="anchor"></a>Visualize cells in reduced dimensions</h2>
<p>The user can visualize the data after quality control in the PCA , t-SNE or UMAP reduced dimensional space. The PCA and t-SNE plots are a convenient way to check if cells form clusters in a way that was expected before any clustering method is applied. For instance, the user should verify whether the QC filtering steps and normalization procedure were efficient by checking the distribution of cells in PC1 and PC2 space. Cells should group independently of normalized coverage.</p>
<p>Depending on wether you have ticked the ‘Run T-SNE’ checkbox in the previous Tab, you will see the T-SNE two dimensional plot.</p>
<p><em>Note:</em> You can change the colors of each samples, batch or cluster and the saved colors will be saved in the Analysis.</p>
<p><br></p>
<div class="figure">
<img src="../inst/www/Guide_images/Step3_UMAP.png" alt=""><p class="caption"><br><br><strong>Visualize Cells: </strong> Visualize cells in reduced dimensions: PCA, UMAP &amp; TSNE. Color cells by sample, batch, or total count to make sure cells are not separating due to library size only.</p>
</div>
<p><br></p>
<p>A brief understanding into what features drives the main PCs is shown in the ‘Contribution to PCA’ box.</p>
<ul>
<li><p>In the barplot, you will get the top loci / closest genes associated to the top loci contributing to the Component of your choice. The top contributing features are taken by sorting the values of the features in the componenent of intereset (e.g. PC1, PC2…). If you observe a huge gap between the top 2-3 features and the rest, this means that very few features are driving most of the dimensionality reduction, and that other less contributing features might be hidden. This also gives you interesting features to look at in the Peak Calling &amp; Coverage or Differential Analysis tabs.</p></li>
<li><p>In the pie chart, the chromosome repartition of the top 100 most contributing features to the componenent of interest (e.g. PC1, PC2…) are displayed. This is used to see if there is any strong imbalance in the “chromosome contribution”, e.g. if a chromosome is contributing more than it should be. Usually, if a single chromosome has &gt; 30% representation, this might indicate that CNVs or genetic events are happening on this chromosome, and you might want to look into more details at these events, or exclude them from the analysis (see Filter &amp; Normalize tab)</p></li>
</ul>
<p><br></p>
<div class="figure">
<img src="../inst/www/Guide_images/Step3_contribution_to_PCA.png" alt=""><p class="caption"><br><br><strong>Most contributing features </strong></p>
</div>
</div>
<div id="cluster-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#cluster-cells" class="anchor"></a>Cluster cells</h2>
<p>Using the first 50 first PCs of computed PCA as input, hierarchical clustering is performed, taking 1-Pearson’s correlation score as distance metric. The cell to cell correlation heatmap and dendogram describing the hierarchical clustering is represented in the left panel.</p>
<p>To improve the stability of the clustering approaches and to remove from the analysis isolated cells that do not belong to any subgroup, cells displaying a Pearson’s pairwise correlation score below a threshold t with at least p % of cells can be filtered out. The correlation threshold t is calculated as a user-defined percentile of Pearson’s pairwise correlation scores for a randomized dataset (percentile is recommended to be set as the 99th percentile). To do so, open the “Filter lowly correlated cells” panel and click on “Filter &amp; Save”. This will filter any cells that are too rare to correlate with enough other cells and form stable clusters.</p>
<p>To select a number of clusters, looking at the correlation heatmap and dendogram, choose a number in the ‘Select Cluster Number’ panel and click on ‘Choose Cluster’.</p>
<p>If you are unsure of how many cluster to choose, you can run consensus clustering using Bioconductor’s package <a href="https://www.bioconductor.org/packages/release/bioc/html/ConsensusClusterPlus.html">ConsensusClusterPlus</a> package to determine what is the appropriate k-partition of the dataset into k clusters. To do so, it evaluates the stability of the clusters and computes item consensus score for each cell for each possible partition from k=2 to n. For each k, consensus partitions of the dataset are done on the basis of 1,000 resampling iterations (80% of cells sampled at each iteration) of hierarchical clustering, with Pearson’s dissimilarity as the distance metric and Ward’s method for linkage analysis. The optimal number of clusters is then chosen by the user; one option is to maximize intra-cluster correlation scores based on the graphics displayed on the ‘Consensus Clustering’ Tab after processing.</p>
<p>If you ran consensus clustering, re-pick the cluster numbers in the ‘Select Cluster Number’ panel making sure you check the ‘Use consensus’ checkbox and re-click on ‘Choose Cluster’.</p>
<p>Cell clusters can then be visualized in two dimensions with the t-SNE or UMAP plot.</p>
<p><em>Note:</em> The correlation filtering and consensus clustering steps are optional as they need long runtime. You might want to proceed to Differential Analysis and Gene Set Analysis directly if you are confident in your choice of clusters.</p>
<p><br></p>
<div class="figure">
<img src="../inst/www/Guide_images/Step4_1_cluster_cells.png" alt=""><p class="caption"><br><br><strong>Cluster Cells: </strong> Choose a number of cluster if you have clear clusters of cells, or run consensus clustering to test multiple clustering and find the optimal cluster number. Optionally filter out lowly correlated cells that might impact the clustering performance.</p>
</div>
<p><br></p>
<p><img src="../inst/www/Guide_images/Step4_UMAP.png" alt="    Cluster Cells - results:  Cluster visualization with UMAP."> ## Inter &amp; Intra correlation violin plots</p>
<hr>
<p>In this box, the heterogeneity of cells in various groups (e.g. sample or cluster) are displayed with two plots:</p>
<ul>
<li><p>In the intra-correlation plot, the distribution of cell-to-cell Pearson correlation scores <strong>within</strong> each cluster or sample is displayed. Heterogeneous groups will have lower correlation than homogeneous groups.</p></li>
<li><p>In the inter-correlation plot, the distribution of cell-to-cell Pearson correlation scores <strong>between</strong> each cluster or sample is displayed. This allows to understand which groups are closer to each other than other groups.</p></li>
</ul>
<p>For both plots, you can add a layer of information by adding the single-cell as jitter points on the plots. This is interesting when combining samples and clusters to see the repartition of conditions into the clusters.</p>
<div class="figure">
<img src="../inst/www/Guide_images/Step4_intracorrelation.png" alt=""><p class="caption"><br><br><strong>Inter/Intra correlation </strong></p>
</div>
</div>
<div id="peak-calling-to-refine-bins-to-gene-annotation" class="section level2">
<h2 class="hasAnchor">
<a href="#peak-calling-to-refine-bins-to-gene-annotation" class="anchor"></a>Peak Calling to refine bins to gene annotation</h2>
<p>This step of the analysis is optional, but recommended for Gene Set Analysis if working on genomic bins in order to refine the peak annotation prior to enrichment analysis. To be able to run this module, some additional command line tools are required such as SAMtools, BEDTools and MACS2. Therefore, this step is only available for Linux and MacOS systems.</p>
<p>You need to input BAM files for the samples (one separate BAM file per sample), with each read being labeled with the barcode ID in the ‘XB’ tag. ChromSCape merges all files and splits them again according to the previously determined clusters of cells (one separate BAM file per cluster). To do so, place all your BAM files in a folder and click on the ‘Choose a folder containing the BAM files’ button to browse for your folder. If some of the BAM files in the folder are not part of your analyis, remove them by pressing the write box and deleting them.</p>
<p>You can select significance threshold for peak detection (passed to MACS2) and merging distance for peaks (defaults to p-value=0.05 and peak merge distance to 5,000). This allows to identify peaks in close proximity (&lt;1,000bp) to a gene transcription start site (TSS); these genes will be later used as input for the enrichment analysis.</p>
<p><br></p>
<div class="figure">
<img src="../inst/www/Guide_images/Step5_peak_calling.png" alt=""><p class="caption"><br><br><strong>Peak Calling to refine bins to gene annotation: </strong> Call peaks on ‘pseudo-bulk’ cell clusters, and refine bins to genes annotation.</p>
</div>
</div>
<div id="differential-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-analysis" class="anchor"></a>Differential Analysis</h2>
<p>To identify differentially enriched regions across single-cells for a given cluster, ChromSCape can perform 1. A non-parametric two-sided Wilcoxon rank sum test comparing normalized counts from individual cells from one cluster versus all other cells, or cluster of choice, 2. Or a parametric test comparing raw counts from individual cells, using the Bioconductor package <a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a>, based on the assumption that the data follows a negative-binomial distribution. We test for the null hypothesis that the distribution of normalized counts from the two compared groups have the same median, with a confidence interval 0.95. The calculated p-values are then corrected by the Benjamini-Hocheberg procedure. You can modify the log2 fold-change threshold and corrected p-value threshold for regions to be considered as significantly differentially enriched (default settings are a p-value and log2 fold-change thresholds respectively of 0.01 and 1).</p>
<p>If you have specified batches to correct for batch effect, the differential analysis is done using the ‘pairwiseWilcox’ function from the Bioconductor package <a href="https://bioconductor.org/packages/release/bioc/html/scran.html">scran</a> , setting the batch of origin as a ‘blocking level’ for each cell.</p>
<div class="figure">
<img src="../inst/www/Guide_images/Step6_DA.png" alt=""><p class="caption"><br><br><strong>Differential Analysis: </strong> Identify most differential features between the various cell clusters</p>
</div>
</div>
<div id="gene-sets-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-sets-analysis" class="anchor"></a>Gene Sets Analysis</h2>
<p>In this tab, using the refined annotation of peaks if available, the final step is to look for enriched gene sets of the MSigDB v5 database within differentially enriched regions (either enriched or depleted regions in the studied histone mark). Click on ‘Start Enrichment Analysis’ in order to apply hypergeometric tests to identify gene sets from the MSigDB v5 database over-represented within differentially enriched regions, correcting for multiple testing with the Benjamini-Hochberg procedure. Users can then visualize most significantly enriched or depleted gene sets corresponding to the epigenetic signatures of each cluster and download gene sets enrichment tables.</p>
<p>The top 100 most significant differential regions, single-cell Epigenomic enrichment levels can be visualized overlaying read counts for each cell at selected genes onto a t-SNE plot (right panel).</p>
<div class="figure">
<img src="../inst/www/Guide_images/Step7_GeneSetAnalysis.png" alt=""><p class="caption"><br><br><strong>Gene Set Analysis: </strong> Look for pathways activated or depleted in most differential regions between clusters</p>
</div>
</div>
</div>
<div id="datasets" class="section level1">
<h1 class="hasAnchor">
<a href="#datasets" class="anchor"></a>Datasets</h1>
<p>Try out ChromSCape with various kind of dataset : <a href="https://www.dropbox.com/sh/vk7umx3ksgoez3x/AACEq9zn-rRbtwf_Al9uEUaQa?dl=0">Dropbox repository</a></p>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h1>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R Under development (unstable) (2021-04-08 r80148)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.5 LTS
## 
## Matrix products: default
## BLAS:   /home/pacome/R/R-devel/lib/libRblas.so
## LAPACK: /home/pacome/R/R-devel/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] ChromSCape_1.1.5 BiocStyle_2.19.2
## 
## loaded via a namespace (and not attached):
##   [1] utf8_1.2.1                   shinydashboard_0.7.1        
##   [3] reticulate_1.20              tidyselect_1.1.1            
##   [5] RSQLite_2.2.7                AnnotationDbi_1.53.2        
##   [7] htmlwidgets_1.5.3            grid_4.1.0                  
##   [9] BiocParallel_1.25.5          Rtsne_0.15                  
##  [11] munsell_0.5.0                ScaledMatrix_0.99.2         
##  [13] ragg_1.1.2                   umap_0.2.7.0                
##  [15] statmod_1.4.36               scran_1.19.23               
##  [17] DT_0.18                      miniUI_0.1.1.1              
##  [19] batchelor_1.7.15             colorspace_2.0-1            
##  [21] Biobase_2.51.0               filelock_1.0.2              
##  [23] knitr_1.33                   rstudioapi_0.13             
##  [25] stats4_4.1.0                 SingleCellExperiment_1.13.14
##  [27] MatrixGenerics_1.3.1         GenomeInfoDbData_1.2.6      
##  [29] bit64_4.0.5                  rprojroot_2.0.2             
##  [31] vctrs_0.3.8                  generics_0.1.0              
##  [33] xfun_0.23                    BiocFileCache_1.99.8        
##  [35] rlist_0.4.6.1                R6_2.5.0                    
##  [37] GenomeInfoDb_1.27.13         ggbeeswarm_0.6.0            
##  [39] rsvd_1.0.5                   locfit_1.5-9.4              
##  [41] msigdbr_7.4.1                bitops_1.0-7                
##  [43] cachem_1.0.5                 DelayedArray_0.17.13        
##  [45] assertthat_0.2.1             promises_1.2.0.1            
##  [47] BiocIO_1.1.2                 shinycssloaders_1.0.0       
##  [49] scales_1.1.1                 beeswarm_0.3.1              
##  [51] gtable_0.3.0                 beachmat_2.7.9              
##  [53] qualV_0.3-3                  rlang_0.4.11                
##  [55] systemfonts_1.0.2            coop_0.6-2                  
##  [57] rtracklayer_1.51.5           lazyeval_0.2.2              
##  [59] BiocManager_1.30.15          yaml_2.2.1                  
##  [61] httpuv_1.6.1                 tools_4.1.0                 
##  [63] bookdown_0.22                ggplot2_3.3.3               
##  [65] ellipsis_0.3.2               kableExtra_1.3.4            
##  [67] jquerylib_0.1.4              BiocGenerics_0.37.6         
##  [69] Rcpp_1.0.6                   sparseMatrixStats_1.3.8     
##  [71] progress_1.2.2               zlibbioc_1.37.0             
##  [73] purrr_0.3.4                  RCurl_1.98-1.3              
##  [75] prettyunits_1.1.1            openssl_1.4.4               
##  [77] viridis_0.6.1                S4Vectors_0.29.19           
##  [79] zoo_1.8-9                    SummarizedExperiment_1.21.3 
##  [81] cluster_2.1.2                colorRamps_2.3              
##  [83] fs_1.5.0                     magrittr_2.0.1              
##  [85] RSpectra_0.16-0              data.table_1.14.0           
##  [87] ResidualMatrix_1.1.0         colourpicker_1.1.0          
##  [89] matrixStats_0.58.0           stringfish_0.15.1           
##  [91] hms_1.1.0                    qs_0.24.1                   
##  [93] shinyjs_2.0.0                mime_0.10                   
##  [95] evaluate_0.14                xtable_1.8-4                
##  [97] XML_3.99-0.6                 IRanges_2.25.11             
##  [99] gridExtra_2.3                compiler_4.1.0              
## [101] biomaRt_2.47.9               shinyhelper_0.3.2           
## [103] scater_1.19.13               tibble_3.1.2                
## [105] KernSmooth_2.23-20           crayon_1.4.1                
## [107] htmltools_0.5.1.1            later_1.2.0                 
## [109] tidyr_1.1.3                  RcppParallel_5.1.4          
## [111] DBI_1.1.1                    RApiSerialize_0.1.0         
## [113] dbplyr_2.1.1                 rappdirs_0.3.3              
## [115] babelgene_21.4               Matrix_1.3-3                
## [117] parallel_4.1.0               metapod_0.99.6              
## [119] igraph_1.2.6                 GenomicRanges_1.43.4        
## [121] forcats_0.5.1                pkgconfig_2.0.3             
## [123] pkgdown_1.6.1                GenomicAlignments_1.27.2    
## [125] plotly_4.9.3                 scuttle_1.1.20              
## [127] xml2_1.3.2                   svglite_2.0.0               
## [129] vipor_0.4.5                  bslib_0.2.5                 
## [131] dqrng_0.3.0                  stringdist_0.9.6.3          
## [133] webshot_0.5.2                XVector_0.31.1              
## [135] rvest_1.0.0                  stringr_1.4.0               
## [137] digest_0.6.27                ConsensusClusterPlus_1.55.0 
## [139] matrixTests_0.1.9            Biostrings_2.59.4           
## [141] rmarkdown_2.8                edgeR_3.33.7                
## [143] DelayedMatrixStats_1.13.6    restfulr_0.0.13             
## [145] curl_4.3.1                   shiny_1.6.0                 
## [147] Rsamtools_2.7.2              rjson_0.2.20                
## [149] lifecycle_1.0.0              jsonlite_1.7.2              
## [151] BiocNeighbors_1.9.4          askpass_1.1                 
## [153] desc_1.3.0                   viridisLite_0.4.0           
## [155] limma_3.47.15                fansi_0.4.2                 
## [157] pillar_1.6.1                 lattice_0.20-44             
## [159] shinyFiles_0.9.0             KEGGREST_1.31.2             
## [161] fastmap_1.1.0                httr_1.4.2                  
## [163] glue_1.4.2                   png_0.1-7                   
## [165] Sushi_1.29.0                 bluster_1.1.12              
## [167] bit_4.0.4                    stringi_1.6.2               
## [169] sass_0.4.0                   blob_1.2.1                  
## [171] textshaping_0.3.4            BiocSingular_1.7.2          
## [173] memoise_2.0.0                dplyr_1.0.6                 
## [175] irlba_2.3.3</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Pacome Prompsy, Celine Vallot.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
