<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Generate a complete ChromSCape analysis — generate_analysis • ChromSCape</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Generate a complete ChromSCape analysis — generate_analysis" />
<meta property="og:description" content="Generate a complete ChromSCape analysis" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ChromSCape</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/vignette.html">Get started</a>
</li>
<li>
  <a href="../articles/vignette.html">Examples</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/vignette.html">ChromSCape overview</a>
    </li>
    <li>
      <a href="../articles/scCutTag_Bartosovic_H3K27me3.html">scCUT&amp;TAG in the mouse brain</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/vallotlab/ChromSCape/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate a complete ChromSCape analysis</h1>
    <small class="dont-index">Source: <a href='https://github.com/vallotlab/ChromSCape/blob/master/R/generate_analysis.R'><code>R/generate_analysis.R</code></a></small>
    <div class="hidden name"><code>generate_analysis.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Generate a complete ChromSCape analysis</p>
    </div>

    <pre class="usage"><span class='fu'>generate_analysis</span><span class='op'>(</span><span class='va'>input_data_folder</span>,
analysis_name <span class='op'>=</span> <span class='st'>"Analysis_1"</span>,
output_directory <span class='op'>=</span> <span class='st'>"./"</span>,
input_data_type <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"scBED"</span>, <span class='st'>"DenseMatrix"</span>, <span class='st'>"SparseMatrix"</span>,
                   <span class='st'>"scBAM"</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span>,
feature_count_on <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"bins"</span>,<span class='st'>"genebody"</span>,<span class='st'>"peaks"</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span>,
feature_count_parameter <span class='op'>=</span> <span class='fl'>50000</span>,
ref_genome <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"hg38"</span>,<span class='st'>"mm10"</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span>,
run <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"filter"</span>, <span class='st'>"CNA"</span>,<span class='st'>"cluster"</span>, <span class='st'>"consensus"</span>,<span class='st'>"peak_call"</span>, <span class='st'>"coverage"</span>, 
       <span class='st'>"DA"</span>, <span class='st'>"GSA"</span>, <span class='st'>"report"</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>,<span class='fl'>3</span>,<span class='fl'>6</span>,<span class='fl'>7</span>,<span class='fl'>8</span>,<span class='fl'>9</span><span class='op'>)</span><span class='op'>]</span>,
min_reads_per_cell <span class='op'>=</span> <span class='fl'>1000</span>,
max_quantile_read_per_cell <span class='op'>=</span> <span class='fl'>99</span>,
n_top_features <span class='op'>=</span> <span class='fl'>40000</span>,
norm_type <span class='op'>=</span> <span class='st'>"CPM"</span>,
subsample_n <span class='op'>=</span> <span class='cn'>NULL</span>,
exclude_regions <span class='op'>=</span> <span class='cn'>NULL</span>,
n_clust <span class='op'>=</span> <span class='cn'>NULL</span>,
corr_threshold <span class='op'>=</span> <span class='fl'>99</span>,
percent_correlation <span class='op'>=</span> <span class='fl'>1</span>,
maxK <span class='op'>=</span> <span class='fl'>10</span>,
qval.th <span class='op'>=</span> <span class='fl'>0.1</span>,
logFC.th <span class='op'>=</span> <span class='fl'>1</span>,
enrichment_qval <span class='op'>=</span> <span class='fl'>0.1</span>,
doBatchCorr  <span class='op'>=</span> <span class='cn'>FALSE</span>,
batch_sels  <span class='op'>=</span> <span class='cn'>NULL</span>,
control_samples_CNA <span class='op'>=</span> <span class='cn'>NULL</span>,
genes_to_plot <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Krt8"</span>,<span class='st'>"Krt5"</span>,<span class='st'>"Tgfb1"</span>, <span class='st'>"Foxq1"</span>, <span class='st'>"Cdkn2b"</span>,
                 <span class='st'>"Cdkn2a"</span>, <span class='st'>"chr7:15000000-20000000"</span><span class='op'>)</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>input_data_folder</th>
      <td><p>Directory containing the input data.</p></td>
    </tr>
    <tr>
      <th>analysis_name</th>
      <td><p>Name given to the analysis.</p></td>
    </tr>
    <tr>
      <th>output_directory</th>
      <td><p>Directory where to create the analysis and the
HTML report.</p></td>
    </tr>
    <tr>
      <th>input_data_type</th>
      <td><p>The type of input data.</p></td>
    </tr>
    <tr>
      <th>feature_count_on</th>
      <td><p>For raw data type, on which features to count the
cells.</p></td>
    </tr>
    <tr>
      <th>feature_count_parameter</th>
      <td><p>Additional parameter corresponding to the
'feature_count_on' parameter. E.g. for 'bins' must be a numeric, e.g. 50000,
for 'peaks' must be a character containing path towards a BED peak file.</p></td>
    </tr>
    <tr>
      <th>ref_genome</th>
      <td><p>The genome of reference.</p></td>
    </tr>
    <tr>
      <th>run</th>
      <td><p>What steps to run. By default runs everything. Some steps are
required in order to run downstream steps.</p></td>
    </tr>
    <tr>
      <th>min_reads_per_cell</th>
      <td><p>Minimum number of reads per cell.</p></td>
    </tr>
    <tr>
      <th>max_quantile_read_per_cell</th>
      <td><p>Upper quantile above which to consider
cells doublets.</p></td>
    </tr>
    <tr>
      <th>n_top_features</th>
      <td><p>Number of features to keep in the analysis.</p></td>
    </tr>
    <tr>
      <th>norm_type</th>
      <td><p>Normalization type.</p></td>
    </tr>
    <tr>
      <th>subsample_n</th>
      <td><p>Number of cells per condition to downsample to, for
performance principally.</p></td>
    </tr>
    <tr>
      <th>exclude_regions</th>
      <td><p>Path towards a BED file containing CNA to exclude from
the analysis (optional).</p></td>
    </tr>
    <tr>
      <th>n_clust</th>
      <td><p>Number of clusters to force choice of clusters.</p></td>
    </tr>
    <tr>
      <th>corr_threshold</th>
      <td><p>Quantile of correlation above which two cells are
considered as correlated.</p></td>
    </tr>
    <tr>
      <th>percent_correlation</th>
      <td><p>Percentage of the total cells that a cell must be
correlated with in order to be kept in the analysis.</p></td>
    </tr>
    <tr>
      <th>maxK</th>
      <td><p>Upper cluster number to rest for ConsensusClusterPlus.</p></td>
    </tr>
    <tr>
      <th>qval.th</th>
      <td><p>Adjusted p-value below which to consider features
differential.</p></td>
    </tr>
    <tr>
      <th>logFC.th</th>
      <td><p>Log2-fold-change above/below which to consider a feature
depleted/enriched.</p></td>
    </tr>
    <tr>
      <th>enrichment_qval</th>
      <td><p>Adjusted p-value below which to consider a gene set as
significantly enriched in differential features.</p></td>
    </tr>
    <tr>
      <th>doBatchCorr</th>
      <td><p>Logical indicating if batch correction using fastMNN
should be run.</p></td>
    </tr>
    <tr>
      <th>batch_sels</th>
      <td><p>If doBatchCorr is TRUE, a named list containing the samples
in each batch.</p></td>
    </tr>
    <tr>
      <th>control_samples_CNA</th>
      <td><p>If running CopyNumber Analysis, a character vector
of the sample names that are 'normal'.</p></td>
    </tr>
    <tr>
      <th>genes_to_plot</th>
      <td><p>A character vector containing genes of interest of which
to plot the coverage.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Creates a ChromSCape-readable directory and saved objects, as well as
a multi-tabbed HTML report resuming the analysis.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='fu'>generate_analysis</span><span class='op'>(</span><span class='st'>"/path/to/data/"</span>, <span class='st'>"Analysis_1"</span><span class='op'>)</span>
<span class='op'>}</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Pacome Prompsy, Celine Vallot.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


