<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Generate cell cluster pseudo-bulk coverage tracks — generate_coverage_tracks • ChromSCape</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Generate cell cluster pseudo-bulk coverage tracks — generate_coverage_tracks" />
<meta property="og:description" content="Generate cell cluster pseudo-bulk coverage tracks. First, scBED
files are concatenated into cell clusters contained in the 'cell_cluster'
column of your SingleCellExperiment object. To do so, for each sample in the
given list, the barcodes of each cluster are grepped and BED files are
merged into pseudo-bulk of clusters (C1,C2...). Two cells from different can
have the same barcode ID as cell affectation is done sample by sample.
Then coverage of pseudo-bulk BED files is calculated by averaging &amp;amp;
smoothing reads on small genomic window (150bp per default). The pseudo
bulk BED and BigWigs coverage tracks are writtend to the output directory.
This functionality is not available on Windows as it uses the 'cat' and
'gzip' utilities from Unix OS." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ChromSCape</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/vignette.html">Get started</a>
</li>
<li>
  <a href="../articles/vignette.html">Examples</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/vignette.html">Vignettes overview</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/vallotlab/ChromSCape/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate cell cluster pseudo-bulk coverage tracks</h1>
    <small class="dont-index">Source: <a href='https://github.com/vallotlab/ChromSCape/blob/master/R/coverage.R'><code>R/coverage.R</code></a></small>
    <div class="hidden name"><code>generate_coverage_tracks.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Generate cell cluster pseudo-bulk coverage tracks. First, scBED
files are concatenated into cell clusters contained in the 'cell_cluster'
column of your SingleCellExperiment object. To do so, for each sample in the
given list, the barcodes of each cluster are grepped and BED files are
merged into pseudo-bulk of clusters (C1,C2...). Two cells from different can
have the same barcode ID as cell affectation is done sample by sample.
Then coverage of pseudo-bulk BED files is calculated by averaging &amp;
smoothing reads on small genomic window (150bp per default). The pseudo
bulk BED and BigWigs coverage tracks are writtend to the output directory.
This functionality is not available on Windows as it uses the 'cat' and
'gzip' utilities from Unix OS.</p>
    </div>

    <pre class="usage"><span class='fu'>generate_coverage_tracks</span><span class='op'>(</span>
  <span class='va'>scExp_cf</span>,
  <span class='va'>input_files_coverage</span>,
  <span class='va'>odir</span>,
  ref_genome <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"hg38"</span>, <span class='st'>"mm10"</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span>,
  bin_width <span class='op'>=</span> <span class='fl'>150</span>,
  n_smoothBin <span class='op'>=</span> <span class='fl'>5</span>,
  read_size <span class='op'>=</span> <span class='fl'>101</span>,
  progress <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>scExp_cf</th>
      <td><p>A SingleCellExperiment with cluster selected.
(see <code><a href='choose_cluster_scExp.html'>choose_cluster_scExp</a></code>). It is recommended having a minimum
of ~100 cells per cluster in order to obtain smooth tracks.</p></td>
    </tr>
    <tr>
      <th>input_files_coverage</th>
      <td><p>A named list of character vector of path towards
single-cell BED files. The names MUST correspond to the 'sample_id' column in
your SingleCellExperiment object. The single-cell BED files names MUST
match the  barcode names in your SingleCellExperiment (column 'barcode'). The
scBED files can be gzipped or not.</p></td>
    </tr>
    <tr>
      <th>odir</th>
      <td><p>The output directory to write the cumulative BED and BigWig
files.</p></td>
    </tr>
    <tr>
      <th>ref_genome</th>
      <td><p>The genome of reference, used to constrain to canonical
chromosomes. Either 'hg38' or 'mm10'. 'hg38' per default.</p></td>
    </tr>
    <tr>
      <th>bin_width</th>
      <td><p>The width of the bin to create the coverage track. The
smaller the greater the resolution &amp; runtime. Default to 150.</p></td>
    </tr>
    <tr>
      <th>n_smoothBin</th>
      <td><p>Number of bins left &amp; right to average ('smooth') the
signal on. Default to 5.</p></td>
    </tr>
    <tr>
      <th>read_size</th>
      <td><p>The estimated size of reads. Default to 101.</p></td>
    </tr>
    <tr>
      <th>progress</th>
      <td><p>A Progress object for Shiny. Default to NULL.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Generate coverage tracks (.bigwig) for each cluster in the
SingleCellExperiment ("cell_cluster" column).</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>scExp</span><span class='op'>)</span>
<span class='va'>input_files_coverage</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>
  <span class='st'>"scChIP_Jurkat_K4me3"</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"/path/to/"</span>,<span class='va'>scExp</span><span class='op'>$</span><span class='va'>barcode</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>51</span><span class='op'>]</span>,<span class='st'>".bed"</span><span class='op'>)</span>,
  <span class='st'>"scChIP_Ramos_K4me3"</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"/path/to/"</span>,<span class='va'>scExp</span><span class='op'>$</span><span class='va'>barcode</span><span class='op'>[</span><span class='fl'>52</span><span class='op'>:</span><span class='fl'>106</span><span class='op'>]</span>,<span class='st'>".bed"</span><span class='op'>)</span>
<span class='op'>)</span>
<span class='fu'>generate_coverage_tracks</span><span class='op'>(</span><span class='va'>scExp</span>, <span class='va'>input_files_coverage</span>, <span class='st'>"/path/to/output"</span>,
ref_genome <span class='op'>=</span> <span class='st'>"hg38"</span><span class='op'>)</span>
<span class='op'>}</span>
</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Pacome Prompsy, Celine Vallot.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


