<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ ChromSCape</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ChromSCape</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/vignette.html">Get started</a>
</li>
<li>
  <a href="../articles/vignette.html">Examples</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/vignette.html">ChromSCape overview</a>
    </li>
    <li>
      <a href="../articles/scCutTag_Bartosovic_H3K27me3.html">scCUT&amp;TAG in the mouse brain</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/vallotlab/ChromSCape/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="CompareWilcox.html">CompareWilcox()</a></code> </p>
        </td>
        <td><p>CompareWilcox</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="CompareedgeRGLM.html">CompareedgeRGLM()</a></code> </p>
        </td>
        <td><p>Creates a summary table with the number of genes under- or overexpressed in
each group and outputs several graphical representations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="DA_custom.html">DA_custom()</a></code> </p>
        </td>
        <td><p>Differential Analysis in 'One vs Rest' mode</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="DA_one_vs_rest_fun.html">DA_one_vs_rest_fun()</a></code> </p>
        </td>
        <td><p>Differential Analysis in 'One vs Rest' mode</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="DA_pairwise.html">DA_pairwise()</a></code> </p>
        </td>
        <td><p>Run differential analysis in Pairwise mode</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="H1proportion.html">H1proportion()</a></code> </p>
        </td>
        <td><p>H1proportion</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="annotToCol2.html">annotToCol2()</a></code> </p>
        </td>
        <td><p>annotToCol2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="annotation_from_merged_peaks.html">annotation_from_merged_peaks()</a></code> </p>
        </td>
        <td><p>Find nearest peaks of each gene and return refined annotation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="anocol_binary.html">anocol_binary()</a></code> </p>
        </td>
        <td><p>Helper binary column for anocol function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="anocol_categorical.html">anocol_categorical()</a></code> </p>
        </td>
        <td><p>Helper binary column for anocol function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bams_to_matrix_indexes.html">bams_to_matrix_indexes()</a></code> </p>
        </td>
        <td><p>Count bam files on interval to create count indexes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="beds_to_matrix_indexes.html">beds_to_matrix_indexes()</a></code> </p>
        </td>
        <td><p>Count bed files on interval to create count indexes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calculate_CNA.html">calculate_CNA()</a></code> </p>
        </td>
        <td><p>Estimate copy number alterations in cytobands</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calculate_cyto_mat.html">calculate_cyto_mat()</a></code> </p>
        </td>
        <td><p>Calculate Fraction of reads in each cytobands</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calculate_gain_or_loss.html">calculate_gain_or_loss()</a></code> </p>
        </td>
        <td><p>Estimate the copy gains/loss of tumor vs normal based on log2-ratio of
fraction of reads</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calculate_logRatio_CNA.html">calculate_logRatio_CNA()</a></code> </p>
        </td>
        <td><p>Calculate the log2-ratio of tumor vs normal fraction of reads in
cytobands</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="call_macs2_merge_peaks.html">call_macs2_merge_peaks()</a></code> </p>
        </td>
        <td><p>Calling MACS2 peak caller and merging resulting peaks</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="changeRange.html">changeRange()</a></code> </p>
        </td>
        <td><p>changeRange</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="check_correct_datamatrix.html">check_correct_datamatrix()</a></code> </p>
        </td>
        <td><p>Check if matrix rownames are well formated and correct if needed</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="choose_cluster_scExp.html">choose_cluster_scExp()</a></code> </p>
        </td>
        <td><p>Choose a number of clusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="choose_perplexity.html">choose_perplexity()</a></code> </p>
        </td>
        <td><p>Choose perplexity depending on number of cells for Tsne</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="col2hex.html">col2hex()</a></code> </p>
        </td>
        <td><p>Col2Hex</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="colors_scExp.html">colors_scExp()</a></code> </p>
        </td>
        <td><p>Adding colors to cells &amp; features</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="combine_datamatrix.html">combine_datamatrix()</a></code> </p>
        </td>
        <td><p>Combine two matrices and emit warning if no regions are in common</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="combine_enrichmentTests.html">combine_enrichmentTests()</a></code> </p>
        </td>
        <td><p>Run enrichment tests and combine into list</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="concatenate_scBed_into_clusters.html">concatenate_scBed_into_clusters()</a></code> </p>
        </td>
        <td><p>Concatenate single-cell BED into clusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="consensus_clustering_scExp.html">consensus_clustering_scExp()</a></code> </p>
        </td>
        <td><p>Wrapper to apply ConsensusClusterPlus to scExp object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="correlation_and_hierarchical_clust_scExp.html">correlation_and_hierarchical_clust_scExp()</a></code> </p>
        </td>
        <td><p>Correlation and hierarchical clustering</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="count_coverage.html">count_coverage()</a></code> </p>
        </td>
        <td><p>Create a smoothed and normalized coverage track from a BAM file and
given a bin GenomicRanges object (same as deepTools bamCoverage)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="create_project_folder.html">create_project_folder()</a></code> </p>
        </td>
        <td><p>Create ChromSCape project folder</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="create_sample_name_mat.html">create_sample_name_mat()</a></code> </p>
        </td>
        <td><p>Create a sample name matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="create_scDataset_raw.html">create_scDataset_raw()</a></code> </p>
        </td>
        <td><p>Create a simulated single cell datamatrix &amp; cell annotation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="create_scExp.html">create_scExp()</a></code> </p>
        </td>
        <td><p>Wrapper to create the single cell experiment from count matrix and feature
dataframe</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="define_feature.html">define_feature()</a></code> </p>
        </td>
        <td><p>Define the features on which reads will be counted</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="detect_samples.html">detect_samples()</a></code> </p>
        </td>
        <td><p>Heuristic discovery of samples based on cell labels</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="differential_analysis_scExp.html">differential_analysis_scExp()</a></code> </p>
        </td>
        <td><p>Runs differential analysis between cell clusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="distPearson.html">distPearson()</a></code> </p>
        </td>
        <td><p>distPearson</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="enrichmentTest.html">enrichmentTest()</a></code> </p>
        </td>
        <td><p>enrichmentTest</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="exclude_features_scExp.html">exclude_features_scExp()</a></code> </p>
        </td>
        <td><p>Remove specific features (CNA, repeats)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="feature_annotation_scExp.html">feature_annotation_scExp()</a></code> </p>
        </td>
        <td><p>Add gene annotations to features</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_correlated_cell_scExp.html">filter_correlated_cell_scExp()</a></code> </p>
        </td>
        <td><p>Filter lowly correlated cells</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_genes_with_refined_peak_annotation.html">filter_genes_with_refined_peak_annotation()</a></code> </p>
        </td>
        <td><p>Filter genes based on peak calling refined annotation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_scExp.html">filter_scExp()</a></code> </p>
        </td>
        <td><p>Filter cells and features</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="find_top_features.html">find_top_features()</a></code> </p>
        </td>
        <td><p>Find most covered features</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_set_enrichment_analysis_scExp.html">gene_set_enrichment_analysis_scExp()</a></code> </p>
        </td>
        <td><p>Runs Gene Set Enrichment Analysis on genes associated with differential
features</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="generate_analysis.html">generate_analysis()</a></code> </p>
        </td>
        <td><p>Generate a complete ChromSCape analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="generate_count_matrix.html">generate_count_matrix()</a></code> </p>
        </td>
        <td><p>Generate count matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="generate_coverage_tracks.html">generate_coverage_tracks()</a></code> </p>
        </td>
        <td><p>Generate cell cluster pseudo-bulk coverage tracks</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="generate_feature_names.html">generate_feature_names()</a></code> </p>
        </td>
        <td><p>Generate feature names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getExperimentNames.html">getExperimentNames()</a></code> </p>
        </td>
        <td><p>Get experiment names from a SingleCellExperiment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getMainExperiment.html">getMainExperiment()</a></code> </p>
        </td>
        <td><p>Get Main experiment of a SingleCellExperiment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_color_dataframe_from_input.html">get_color_dataframe_from_input()</a></code> </p>
        </td>
        <td><p>Get color dataframe from shiny::colorInput</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_cyto_features.html">get_cyto_features()</a></code> </p>
        </td>
        <td><p>Map features onto cytobands</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_genomic_coordinates.html">get_genomic_coordinates()</a></code> </p>
        </td>
        <td><p>Get SingleCellExperiment's genomic coordinates</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_most_variable_cyto.html">get_most_variable_cyto()</a></code> </p>
        </td>
        <td><p>Retrieve the cytobands with the most variable fraction of reads</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gg_fill_hue.html">gg_fill_hue()</a></code> </p>
        </td>
        <td><p>gg_fill_hue</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="groupMat.html">groupMat()</a></code> </p>
        </td>
        <td><p>groupMat</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="has_genomic_coordinates.html">has_genomic_coordinates()</a></code> </p>
        </td>
        <td><p>Does SingleCellExperiment has genomic coordinates in features ?</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="hclustAnnotHeatmapPlot.html">hclustAnnotHeatmapPlot()</a></code> </p>
        </td>
        <td><p>hclustAnnotHeatmapPlot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="hg38.GeneTSS.html">hg38.GeneTSS</a></code> </p>
        </td>
        <td><p>Data.frame of gene TSS - hg38</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="hg38.chromosomes.html">hg38.chromosomes</a></code> </p>
        </td>
        <td><p>Data.frame of chromosome length - hg38</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="hg38.cytoBand.html">hg38.cytoBand</a></code> </p>
        </td>
        <td><p>Data.frame of cytoBandlocation - hg38</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="imageCol.html">imageCol()</a></code> </p>
        </td>
        <td><p>imageCol</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="import_count_input_files.html">import_count_input_files()</a></code> </p>
        </td>
        <td><p>Import and count input files depending on their format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="import_scExp.html">import_scExp()</a></code> </p>
        </td>
        <td><p>Read single-cell matrix(ces) into scExp</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="index_peaks_barcodes_to_matrix_indexes.html">index_peaks_barcodes_to_matrix_indexes()</a></code> </p>
        </td>
        <td><p>Read index-peaks-barcodes trio files on interval to create count indexes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="inter_correlation_scExp.html">inter_correlation_scExp()</a></code> </p>
        </td>
        <td><p>Calculate inter correlation between cluster or samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="intra_correlation_scExp.html">intra_correlation_scExp()</a></code> </p>
        </td>
        <td><p>Calculate intra correlation between cluster or samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="launchApp.html">launchApp()</a></code> </p>
        </td>
        <td><p>Launch ChromSCape</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="load_MSIGdb.html">load_MSIGdb()</a></code> </p>
        </td>
        <td><p>Load and format MSIGdb pathways using msigdbr package</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="merge_MACS2_peaks.html">merge_MACS2_peaks()</a></code> </p>
        </td>
        <td><p>Merge peak files from MACS2 peak caller</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mm10.GeneTSS.html">mm10.GeneTSS</a></code> </p>
        </td>
        <td><p>Data.frame of gene TSS - mm10</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mm10.chromosomes.html">mm10.chromosomes</a></code> </p>
        </td>
        <td><p>Data.frame of chromosome length - mm10</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mm10.cytoBand.html">mm10.cytoBand</a></code> </p>
        </td>
        <td><p>Data.frame of cytoBandlocation - mm10</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="normalize_scExp.html">normalize_scExp()</a></code> </p>
        </td>
        <td><p>Normalize counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="num_cell_after_QC_filt_scExp.html">num_cell_after_QC_filt_scExp()</a></code> </p>
        </td>
        <td><p>Table of cells before / after QC</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="num_cell_after_cor_filt_scExp.html">num_cell_after_cor_filt_scExp()</a></code> </p>
        </td>
        <td><p>Number of cells before &amp; after correlation filtering</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="num_cell_before_cor_filt_scExp.html">num_cell_before_cor_filt_scExp()</a></code> </p>
        </td>
        <td><p>Table of number of cells before correlation filtering</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="num_cell_in_cluster_scExp.html">num_cell_in_cluster_scExp()</a></code> </p>
        </td>
        <td><p>Number of cells in each cluster</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="num_cell_scExp.html">num_cell_scExp()</a></code> </p>
        </td>
        <td><p>Table of cells</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pca_irlba_for_sparseMatrix.html">pca_irlba_for_sparseMatrix()</a></code> </p>
        </td>
        <td><p>Run sparse PCA using irlba SVD</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="peaks_to_bins.html">peaks_to_bins()</a></code> </p>
        </td>
        <td><p>Transforms a peaks x cells count matrix into a bins x cells count matrix.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_cluster_consensus_scExp.html">plot_cluster_consensus_scExp()</a></code> </p>
        </td>
        <td><p>Plot cluster consensus</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_coverage_BigWig.html">plot_coverage_BigWig()</a></code> </p>
        </td>
        <td><p>Coverage plot using Sushi</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_differential_H1_scExp.html">plot_differential_H1_scExp()</a></code> </p>
        </td>
        <td><p>Differential H1 distribution plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_differential_summary_scExp.html">plot_differential_summary_scExp()</a></code> </p>
        </td>
        <td><p>Differential summary barplot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_differential_volcano_scExp.html">plot_differential_volcano_scExp()</a></code> </p>
        </td>
        <td><p>Volcano plot of differential features</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_distribution_scExp.html">plot_distribution_scExp()</a></code> </p>
        </td>
        <td><p>Plotting distribution of signal</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_gain_or_loss_barplots.html">plot_gain_or_loss_barplots()</a></code> <code><a href="plot_gain_or_loss_barplots.html">plot_gain_or_loss_barplots()</a></code> </p>
        </td>
        <td><p>Plot Gain or Loss of cytobands of the most variables cytobands</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_heatmap_scExp.html">plot_heatmap_scExp()</a></code> </p>
        </td>
        <td><p>Plot cell correlation heatmap with annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_inter_correlation_scExp.html">plot_inter_correlation_scExp()</a></code> </p>
        </td>
        <td><p>Violin plot of inter-correlation distribution between one or multiple groups
and one reference group</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_intra_correlation_scExp.html">plot_intra_correlation_scExp()</a></code> </p>
        </td>
        <td><p>Violin plot of intra-correlation distribution</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_most_contributing_features.html">plot_most_contributing_features()</a></code> </p>
        </td>
        <td><p>Plot Top/Bottom most contributing features to PCA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_pie_most_contributing_chr.html">plot_pie_most_contributing_chr()</a></code> </p>
        </td>
        <td><p>Pie chart of top contribution of chromosomes in the 100 most contributing
features to PCA
#'</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_reduced_dim_scExp.html">plot_reduced_dim_scExp()</a></code> </p>
        </td>
        <td><p>Plot reduced dimensions (PCA, TSNE, UMAP)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_reduced_dim_scExp_CNA.html">plot_reduced_dim_scExp_CNA()</a></code> </p>
        </td>
        <td><p>Plot UMAP colored by Gain or Loss of cytobands</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="preprocess_CPM.html">preprocess_CPM()</a></code> </p>
        </td>
        <td><p>Preprocess scExp - Counts Per Million (CPM)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="preprocess_RPKM.html">preprocess_RPKM()</a></code> </p>
        </td>
        <td><p>Preprocess scExp - Read per Kilobase Per Million (RPKM)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="preprocess_TFIDF.html">preprocess_TFIDF()</a></code> </p>
        </td>
        <td><p>Preprocess scExp - TF-IDF</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="preprocess_TPM.html">preprocess_TPM()</a></code> </p>
        </td>
        <td><p>Preprocess scExp - Transcripts per Million (TPM)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="preprocess_feature_size_only.html">preprocess_feature_size_only()</a></code> </p>
        </td>
        <td><p>Preprocess scExp - size only</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="raw_counts_to_sparse_matrix.html">raw_counts_to_sparse_matrix()</a></code> </p>
        </td>
        <td><p>Create a sparse count matrix from various format of input data.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rawfile_ToBigWig.html">rawfile_ToBigWig()</a></code> </p>
        </td>
        <td><p>rawfile_ToBigWig : reads in BAM file and write out BigWig coverage file,
normalized and smoothed</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_count_mat_with_separated_chr_start_end.html">read_count_mat_with_separated_chr_start_end()</a></code> </p>
        </td>
        <td><p>Read a count matrix with three first columns (chr,start,end)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_sparse_matrix.html">read_sparse_matrix()</a></code> </p>
        </td>
        <td><p>Read in one or multiple sparse matrices (10X format)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reduce_dim_batch_correction.html">reduce_dim_batch_correction()</a></code> </p>
        </td>
        <td><p>Reduce dimension with batch corrections</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reduce_dims_scExp.html">reduce_dims_scExp()</a></code> </p>
        </td>
        <td><p>Reduce dimensions (PCA, TSNE, UMAP)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="remove_chr_M_fun.html">remove_chr_M_fun()</a></code> </p>
        </td>
        <td><p>Remove chromosome M from scExprownames</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="remove_non_canonical_fun.html">remove_non_canonical_fun()</a></code> </p>
        </td>
        <td><p>Remove non canonical chromosomes from scExp</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="results_enrichmentTest.html">results_enrichmentTest()</a></code> </p>
        </td>
        <td><p>Resutls of hypergeometric gene set enrichment test</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="retrieve_top_bot_features_pca.html">retrieve_top_bot_features_pca()</a></code> </p>
        </td>
        <td><p>Retrieve Top and Bot most contributing features of PCA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_pairwise_tests.html">run_pairwise_tests()</a></code> </p>
        </td>
        <td><p>Run pairwise tests</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_tsne_scExp.html">run_tsne_scExp()</a></code> </p>
        </td>
        <td><p>Run tsne on single cell experiment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scExp.html">scExp</a></code> </p>
        </td>
        <td><p>A SingleCellExperiment outputed by ChromSCape</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="separate_BAM_into_clusters.html">separate_BAM_into_clusters()</a></code> </p>
        </td>
        <td><p>Separate BAM files into cell cluster BAM files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="separator_count_mat.html">separator_count_mat()</a></code> </p>
        </td>
        <td><p>Determine Count matrix separator ("tab" or ",")</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smoothBin.html">smoothBin()</a></code> </p>
        </td>
        <td><p>Smooth a vector of values with nb_bins left and righ values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subsample_scExp.html">subsample_scExp()</a></code> </p>
        </td>
        <td><p>Subsample scExp</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subset_bam_call_peaks.html">subset_bam_call_peaks()</a></code> </p>
        </td>
        <td><p>Peak calling on cell clusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="swapAltExp_sameColData.html">swapAltExp_sameColData()</a></code> </p>
        </td>
        <td><p>Swap main &amp; alternative Experiments, with fixed colData</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="table_enriched_genes_scExp.html">table_enriched_genes_scExp()</a></code> </p>
        </td>
        <td><p>Creates table of enriched genes sets</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="warning_DA.html">warning_DA()</a></code> </p>
        </td>
        <td><p>Warning for differential_analysis_scExp</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="warning_filter_correlated_cell_scExp.html">warning_filter_correlated_cell_scExp()</a></code> </p>
        </td>
        <td><p>warning_filter_correlated_cell_scExp</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="warning_plot_reduced_dim_scExp.html">warning_plot_reduced_dim_scExp()</a></code> </p>
        </td>
        <td><p>A warning helper for plot_reduced_dim_scExp</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="warning_raw_counts_to_sparse_matrix.html">warning_raw_counts_to_sparse_matrix()</a></code> </p>
        </td>
        <td><p>Warning for raw_counts_to_sparse_matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="wrapper_Signac_FeatureMatrix.html">wrapper_Signac_FeatureMatrix()</a></code> </p>
        </td>
        <td><p>Wrapper around 'FeatureMatrix' function from Signac Package</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Pacome Prompsy, Celine Vallot.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


