<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Peak calling on cell clusters — subset_bam_call_peaks • ChromSCape</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Peak calling on cell clusters — subset_bam_call_peaks" />
<meta property="og:description" content="This functions does peak calling on each cell population in order to refine
gene annotation for large bins. For instance, a 50000bp bins might contain
the TSS of several genes, while in reality only one or two of these genes are
overlapping the signal (peak). To do so, first in-silico cell sorting is
applied based on previously defined clusters contained in the
SingleCellExperiment. Taking BAM files of each sample as input, samtools
pools then splits reads from each cell barcode into 1 BAM file per cell
cluster (pseudo-bulk). Then MACS2 calls peaks on each cluster. The peaks are
aggregated and merged if closer to a certain distance
defined by user (10000bp). Then,
This function takes as input a SingleCellExperiment, that must contain a
'cell_cluster' column in it's colData, an output directory where to store
temporary files, the list of BAM files corresponding to each sample and
containing the cell barcode information as a tag (for instance tag CB:Z:xxx,
XB:Z:xxx or else...) or single-cell BED files containing the raw reads and
corresponding to the 'barcode' column metadata,
the p.value used by MACS2 to distinguish significant
peaks, the reference genome (either hg38 or mm10), the maximal merging
distance in bp and a data.frame containing gene TSS genomic cooridnates of
corresponding genome (if set to NULL, will automatically load geneTSS). The
output is a SingleCellExperiment with GRanges object containing ranges of
each merged peaks that falls within genomic bins of the SingleCellExperiment,
saving the bin range as additional column (window_chr, window_start,
window_end), as well as the closests genes and their distance relative to the
peak. The peaks may be present in several rows if multiple genes are close /
overlap to the peaks.
Note that the user must have MACS2 installed and available in the
PATH. Users can open command terminal and type 'which
macs2' to verify the availability of these programs. Will only work on unix
operating system. Check operating system with 'print(.Platform)'." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ChromSCape</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/vignette.html">Get started</a>
</li>
<li>
  <a href="../articles/vignette.html">Examples</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/vignette.html">Vignettes overview</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/vallotlab/ChromSCape/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Peak calling on cell clusters</h1>
    <small class="dont-index">Source: <a href='https://github.com/vallotlab/ChromSCape/blob/master/R/peak_calling.R'><code>R/peak_calling.R</code></a></small>
    <div class="hidden name"><code>subset_bam_call_peaks.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This functions does peak calling on each cell population in order to refine
gene annotation for large bins. For instance, a 50000bp bins might contain
the TSS of several genes, while in reality only one or two of these genes are
overlapping the signal (peak). To do so, first in-silico cell sorting is
applied based on previously defined clusters contained in the
SingleCellExperiment. Taking BAM files of each sample as input, samtools
pools then splits reads from each cell barcode into 1 BAM file per cell
cluster (pseudo-bulk). Then MACS2 calls peaks on each cluster. The peaks are
aggregated and merged if closer to a certain distance
defined by user (10000bp). Then,</p>
<p>This function takes as input a SingleCellExperiment, that must contain a
'cell_cluster' column in it's colData, an output directory where to store
temporary files, the list of BAM files corresponding to each sample and
containing the cell barcode information as a tag (for instance tag CB:Z:xxx,
XB:Z:xxx or else...) or single-cell BED files containing the raw reads and
corresponding to the 'barcode' column metadata,
the p.value used by MACS2 to distinguish significant
peaks, the reference genome (either hg38 or mm10), the maximal merging
distance in bp and a data.frame containing gene TSS genomic cooridnates of
corresponding genome (if set to NULL, will automatically load geneTSS). The
output is a SingleCellExperiment with GRanges object containing ranges of
each merged peaks that falls within genomic bins of the SingleCellExperiment,
saving the bin range as additional column (window_chr, window_start,
window_end), as well as the closests genes and their distance relative to the
peak. The peaks may be present in several rows if multiple genes are close /
overlap to the peaks.</p>
<p>Note that the user must have MACS2 installed and available in the
PATH. Users can open command terminal and type 'which
macs2' to verify the availability of these programs. Will only work on unix
operating system. Check operating system with 'print(.Platform)'.</p>
    </div>

    <pre class="usage"><span class='fu'>subset_bam_call_peaks</span><span class='op'>(</span>
  <span class='va'>scExp</span>,
  <span class='va'>odir</span>,
  <span class='va'>input</span>,
  format <span class='op'>=</span> <span class='st'>"BAM"</span>,
  p.value <span class='op'>=</span> <span class='fl'>0.05</span>,
  ref <span class='op'>=</span> <span class='st'>"hg38"</span>,
  peak_distance_to_merge <span class='op'>=</span> <span class='fl'>10000</span>,
  geneTSS_annotation <span class='op'>=</span> <span class='cn'>NULL</span>,
  run_coverage <span class='op'>=</span> <span class='cn'>FALSE</span>,
  progress <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>scExp</th>
      <td><p>A SingleCellExperiment object</p></td>
    </tr>
    <tr>
      <th>odir</th>
      <td><p>Output directory where to write temporary files and each
cluster's BAM file</p></td>
    </tr>
    <tr>
      <th>input</th>
      <td><p>A character vector of file paths to each sample's BAM file,
containing cell barcode information as tags. BAM files can be paired-end or
single-end.</p></td>
    </tr>
    <tr>
      <th>format</th>
      <td><p>Format of the input data, either "BAM" or "scBED".</p></td>
    </tr>
    <tr>
      <th>p.value</th>
      <td><p>a p-value to use for MACS2 to determine significant peaks.
(0.05)</p></td>
    </tr>
    <tr>
      <th>ref</th>
      <td><p>A reference genome, either hg38 or mm10. ('hg38')</p></td>
    </tr>
    <tr>
      <th>peak_distance_to_merge</th>
      <td><p>Maximal distance to merge peaks together after
peak calling , in bp. (10000)</p></td>
    </tr>
    <tr>
      <th>geneTSS_annotation</th>
      <td><p>A data.frame annotation of genes TSS. If NULL will
automatically load Gencode list of genes fro specified reference genome.</p></td>
    </tr>
    <tr>
      <th>run_coverage</th>
      <td><p>Create coverage tracks (.bw) for each cluster ?</p></td>
    </tr>
    <tr>
      <th>progress</th>
      <td><p>A shiny Progress instance to display progress bar.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A SingleCellExperiment with refinded annotation</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The BED files of the peaks called for each clusters, as well as
the merged peaks are written in the output directory.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"scExp"</span><span class='op'>)</span>
<span class='fu'>subset_bam_call_peaks</span><span class='op'>(</span><span class='va'>scExp</span>, <span class='st'>"path/to/out/"</span>, <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='st'>"sample1"</span> <span class='op'>=</span> 
   <span class='st'>"path/to/BAM/sample1.bam"</span>, <span class='st'>"sample2"</span> <span class='op'>=</span> <span class='st'>"path/to/BAM/sample2.bam"</span><span class='op'>)</span>,
   p.value <span class='op'>=</span> <span class='fl'>0.05</span>, ref <span class='op'>=</span> <span class='st'>"hg38"</span>, peak_distance_to_merge <span class='op'>=</span> <span class='fl'>10000</span>, 
   geneTSS_annotation <span class='op'>=</span> <span class='cn'>NULL</span><span class='op'>)</span>
<span class='op'>}</span>
</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Pacome Prompsy, Celine Vallot.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


